% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/import_utils.R
\name{prepare_data}
\alias{prepare_data}
\title{Prepare data for use in PLN models}
\usage{
prepare_data(counts, covariates, offset = "TSS", ...)
}
\arguments{
\item{counts}{Required. An abundance count table, preferably with dimensions names and species as columns.}

\item{covariates}{Required. A covariates data frame, preferably with row names.}

\item{offset}{Optional. Normalisation scheme used to compute scaling factors used as offset during PLN inference. Can be "TSS" (Total Sum Scaling, default), "CSS" (Cumulative Sum Scaling, used in metagenomeSeq), "RLE" (Relative Log Expression, used in DESeq2), "GMPR" (Geometric Mean of Pairwise Ratio, introduced in Chen et al., 2018) or "none".}

\item{...}{Additional parameters passed on to \code{\link[=compute_offset]{compute_offset}}}
}
\value{
A data.frame suited for use in \code{\link[=PLN]{PLN}} and its variants with two specials components: an abundance count matrix (in component "Abundance") and an offset vector/matrix (in component "Offset", only if offset is not set to "none")
}
\description{
Prepare data in proper format for use in PLN model and its variants. The function (i) merges a count table and
a covariate data frame in the most comprehensive way and (ii) computes offsets from the count table using one of several normalization schemes (TSS, CSS, RLE, GMPR, etc). The function fails with informative messages when the heuristics used for sample matching fail.
}
\examples{
data(trichoptera)
proper_data <- prepare_data(
 counts      = trichoptera$Abundance,
 covariates = trichoptera$Covariate,
  offset = "TSS"
)
proper_data$Abundance
proper_data$Offset
}
\references{
Chen, L., Reeve, J., Zhang, L., Huang, S., Wang, X. and Chen, J. (2018) GMPR: A robust normalization method for zero-inflated count data with application to microbiome sequencing data. PeerJ, 6, e4600 \url{https://doi.org/10.7717/peerj.4600}

Paulson, J. N., Colin Stine, O., Bravo, H. C. and Pop, M. (2013) Differential abundance analysis for microbial marker-gene surveys. Nature Methods, 10, 1200-1202 \url{http://dx.doi.org/10.1038/nmeth.2658}

Anders, S. and Huber, W. (2010) Differential expression analysis for sequence count data. Genome Biology, 11, R106 \url{https://doi.org/10.1186/gb-2010-11-10-r106}
}
\seealso{
\code{\link[=compute_offset]{compute_offset}} for details on the different normalisation schemes
}
