<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Prepare data for use in PLN models — prepare_data • PLNmodels</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Prepare data for use in PLN models — prepare_data" />
<meta property="og:description" content="Prepare data in proper format for use in PLN model and its variants. The function (i) merges a count table and
a covariate data frame in the most comprehensive way and (ii) computes offsets from the count table using one of several normalization schemes (TSS, CSS, RLE, GMPR, Wrench, etc). The function fails with informative messages when the heuristics used for sample matching fail." />
<meta property="og:image" content="https://pln-team.github.io/PLNmodels/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PLNmodels</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.5-9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://pln-team.github.io/slideshow/slides">Slideshow on PLN</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Trichoptera.html">The trichoptera data set</a>
    </li>
    <li>
      <a href="../articles/Import_data.html">Data importation in PLNmodels</a>
    </li>
    <li>
      <a href="../articles/PLN.html">PLN: Multivariate Poisson regression</a>
    </li>
    <li>
      <a href="../articles/PLNPCA.html">PLNPCA: Dimension Reduction</a>
    </li>
    <li>
      <a href="../articles/PLNnetwork.html">PLNnetwork: sparse structure inference</a>
    </li>
    <li>
      <a href="../articles/PLNLDA.html">PLNLDA: Discriminant Analysis</a>
    </li>
    <li>
      <a href="../articles/PLNmixture.html">PLNmixture: Model-based clustering for counts</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/pln-team/PLNmodels/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Prepare data for use in PLN models</h1>
    <small class="dont-index">Source: <a href='https://github.com/pln-team/PLNmodels/blob/master/R/import_utils.R'><code>R/import_utils.R</code></a></small>
    <div class="hidden name"><code>prepare_data.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Prepare data in proper format for use in PLN model and its variants. The function (i) merges a count table and
a covariate data frame in the most comprehensive way and (ii) computes offsets from the count table using one of several normalization schemes (TSS, CSS, RLE, GMPR, Wrench, etc). The function fails with informative messages when the heuristics used for sample matching fail.</p>
    </div>

    <pre class="usage"><span class='fu'>prepare_data</span><span class='op'>(</span><span class='va'>counts</span>, <span class='va'>covariates</span>, offset <span class='op'>=</span> <span class='st'>"TSS"</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>counts</th>
      <td><p>Required. An abundance count table, preferably with dimensions names and species as columns.</p></td>
    </tr>
    <tr>
      <th>covariates</th>
      <td><p>Required. A covariates data frame, preferably with row names.</p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>Optional. Normalization scheme used to compute scaling factors used as offset during PLN inference. Available schemes are "TSS" (Total Sum Scaling, default), "CSS" (Cumulative Sum Scaling, used in metagenomeSeq), "RLE" (Relative Log Expression, used in DESeq2), "GMPR" (Geometric Mean of Pairwise Ratio, introduced in Chen et al., 2018), Wrench (introduced in Kumar et al., 2018) or "none". Alternatively the user can supply its own vector or matrix of offsets (see note for specification of the user-supplied offsets).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional parameters passed on to <code><a href='compute_offset.html'>compute_offset()</a></code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data.frame suited for use in <code><a href='PLN.html'>PLN()</a></code> and its variants with two specials components: an abundance count matrix (in component "Abundance") and an offset vector/matrix (in component "Offset", only if offset is not set to "none")</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>User supplied offsets should be either vectors/column-matrices or have the same number of column as the original count matrix and either (i) dimension names or (ii) the same dimensions as the count matrix. Samples are trimmed in exactly the same way to remove empty samples.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Chen, L., Reeve, J., Zhang, L., Huang, S., Wang, X. and Chen, J. (2018) GMPR: A robust normalization method for zero-inflated count data with application to microbiome sequencing data. PeerJ, 6, e4600 doi: <a href='https://doi.org/10.7717/peerj.4600'>10.7717/peerj.4600</a></p>
<p>Paulson, J. N., Colin Stine, O., Bravo, H. C. and Pop, M. (2013) Differential abundance analysis for microbial marker-gene surveys. Nature Methods, 10, 1200-1202 doi: <a href='https://doi.org/10.1038/nmeth.2658'>10.1038/nmeth.2658</a></p>
<p>Anders, S. and Huber, W. (2010) Differential expression analysis for sequence count data. Genome Biology, 11, R106 doi: <a href='https://doi.org/10.1186/gb-2010-11-10-r106'>10.1186/gb-2010-11-10-r106</a></p>
<p>Kumar, M., Slud, E., Okrah, K. et al. (2018) Analysis and correction of compositional bias in sparse sequencing count data. BMC Genomics 19, 799 doi: <a href='https://doi.org/10.1186/s12864-018-5160-5'>10.1186/s12864-018-5160-5</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='compute_offset.html'>compute_offset()</a></code> for details on the different normalization schemes</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>trichoptera</span><span class='op'>)</span>
<span class='va'>proper_data</span> <span class='op'>&lt;-</span> <span class='fu'>prepare_data</span><span class='op'>(</span>
 counts     <span class='op'>=</span> <span class='va'>trichoptera</span><span class='op'>$</span><span class='va'>Abundance</span>,
 covariates <span class='op'>=</span> <span class='va'>trichoptera</span><span class='op'>$</span><span class='va'>Covariate</span>,
 offset     <span class='op'>=</span> <span class='st'>"TSS"</span>
<span class='op'>)</span>
<span class='va'>proper_data</span><span class='op'>$</span><span class='va'>Abundance</span>
</div><div class='output co'>#&gt;    Che Hyc Hym Hys  Psy Aga Glo Ath Cea Ced Set All Han Hfo Hsp Hve Sta
#&gt; 1    0   0   5   0   17   0   0   0   0   2   0   1   0   1   2   0   1
#&gt; 2    0   0   3   0    8   0   0   0   0   0   0   2   0   0   0   0   0
#&gt; 3    0   0   1   0   32   0   0   0   0   0   0   4   0   0   1   0   0
#&gt; 4    0   0   3   0  176   4   0   0   0   1   0   3   0   0   3   0   2
#&gt; 5    0   0   4   0   69   2   0   0   0   0   0   1   0   0   1   0   2
#&gt; 6    0   0   2   0   14   1   0   0   0   0   0   1   0   0   0   0   0
#&gt; 7    0   0   2   0    4   0   0   0   0   2   0   0   0   0   0   0   0
#&gt; 8    0   0   1   0   29   1   0   0   0   0   0   0   0   0   0   0   3
#&gt; 9    0   0   4   0    8   0   0   0   0   0   0   0   0   0   0   0   0
#&gt; 10   0   0   2   0    2   0   0   0   0   0   0   0   0   0   0   0   0
#&gt; 11   0   0   1   0    3   0   0   0   0   0   0   0   0   0   0   0   0
#&gt; 12   0   0   0   0    3   0   0   0   0   0   0   0   0   0   0   0   0
#&gt; 13   1   0   1   0   40   2   0   1   0   2   0   1   0   0   0   1   0
#&gt; 14   0   0   4   0   20   2   1   0   0   1   0   0   0   0   1   0   4
#&gt; 15   0   0   3   0  500  19   0   2   0  17   5   2   0   0   2   1  49
#&gt; 16   0   0   2   0  142   9   0   0   0   3   3   2   0   0   0   0  11
#&gt; 17   0   0   2   0   44   1   0   0   1   4   3   0   0   0   0   0   3
#&gt; 18   0   0   3   1   31   4   0   0   0   1   7   0   0   0   0   0   4
#&gt; 19   0   0   6   0   30   6   0   0   0   2   5   0   0   0   0   0   7
#&gt; 20   0   1   7   0   71   5   1   1   0   5  18   0   0   1   5   0  12
#&gt; 21   0   0   5   0   20   0   0   0   0   5   1   0   0   0   0   0   4
#&gt; 22   0   0   3   0    5   0   0   0   0   2   0   0   0   0   2   0   1
#&gt; 23   0   0   0   1    8   0   0   0   0   1   0   0   0   1   4   0   2
#&gt; 24   0   0   1   0    2   0   0   0   0   0   0   0   0   0   0   0   0
#&gt; 25   0   0   3   0   10   1   1   0   0   0   2   0   0   2   7   0   1
#&gt; 26   0   0   1   0    9   0   0   0   0   1   2   1   3   3  14   0   6
#&gt; 27   0   0   1   0   14   0   0   0   0   5   3   0   7   0  10   0   4
#&gt; 28   0   0   0   0    2   0   0   1   0   0   0   0   0   1   3   0   1
#&gt; 29   0   0   0   0    4   0   0   0   0   0   1   0   0   0   3   0   1
#&gt; 30   0   1   0   0 1173   2   3   1   2   9  71   0  27  69 227   1  13
#&gt; 31   0   0  12   0 2671   1   4   1   3   7  49   1  15  52 161   1   2
#&gt; 32   0   0   7   0   33   4   0   0   0   0   0   3  33   1   0   0   7
#&gt; 33   0   0  12   0   62   8   0   0   0   3   1  15  23   0   5   0   6
#&gt; 34   0   0  15   0  220  20   0   0   0   2   0   5  27   0   5   0  33
#&gt; 35   0   0   6   0   29   6   0   0   0   3   0   2   3   0   0   0  17
#&gt; 36   0   0   4   0   60   3   0   0   0   0   0   0   2   0   0   0  21
#&gt; 37   0   0  18   1   21   2   0   0   0   2   1   2   5   0   0   0  11
#&gt; 38   0   0   4   0    3   0   0   0   0   4   0   0   1   0   0   0   3
#&gt; 39   0   0   5   0    5   0   0   0   0   1   0   0   1   0   0   0   2
#&gt; 40   0   0   1   0   18   0   0   0   0   0   0   1   0   0   0   0   0
#&gt; 41   0   0   4   0   49   0   0   0   0   3   0   1   0   0   0   2  11
#&gt; 42   0   0   3   1   33   0   0   2   0   1   0   0   2   0   0   0  11
#&gt; 43   0   0   3   0   71   1   0   0   0  11   1   0   3   0   0   0   5
#&gt; 44   0   0   6   1   28   1   2   1   0   3   0   1   0   0   0   0   0
#&gt; 45   0   0   5   1   37   3   1   2   0   7   1   1   0   0   0   1   3
#&gt; 46   1   0   4   0  103   1   1   0   0   2  10   2   0   2   3   2  18
#&gt; 47   0   0   2   0   11   0   0   1   0   1   1   0   0   0   0   0   0
#&gt; 48   1   0   0   0   17   0   0   1   1   2   0   0   3   0   0   0   6
#&gt; 49   0   1   2   0   27   0   0   0   0   1   4   0  36   0  11   0   4</div><div class='input'><span class='va'>proper_data</span><span class='op'>$</span><span class='va'>Offset</span>
</div><div class='output co'>#&gt;  [1]   29   13   38  192   79   18    8   34   12    4    4    3   49   33  600
#&gt; [16]  172   58   51   56  127   35   13   17    3   27   40   44    8    9 1599
#&gt; [31] 2980   88  135  327   66   90   63   15   14   20   70   53   95   43   62
#&gt; [46]  149   16   31   86</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Julien Chiquet, Mahendra Mariadassou, François Gindraud.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


