<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data importation in PLNmodels • PLNmodels</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data importation in PLNmodels">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PLNmodels</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.1-9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://pln-team.github.io/slideshow/slides" class="external-link">Slideshow on PLN</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Trichoptera.html">The trichoptera data set</a>
    </li>
    <li>
      <a href="../articles/Import_data.html">Data importation in PLNmodels</a>
    </li>
    <li>
      <a href="../articles/PLN.html">PLN: Multivariate Poisson regression</a>
    </li>
    <li>
      <a href="../articles/PLNPCA.html">PLNPCA: Dimension Reduction</a>
    </li>
    <li>
      <a href="../articles/PLNnetwork.html">PLNnetwork: sparse structure inference</a>
    </li>
    <li>
      <a href="../articles/PLNLDA.html">PLNLDA: Discriminant Analysis</a>
    </li>
    <li>
      <a href="../articles/PLNmixture.html">PLNmixture: Model-based clustering for counts</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pln-team/PLNmodels/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data importation in PLNmodels</h1>
                        <h4 data-toc-skip class="author">PLN team</h4>
            
            <h4 data-toc-skip class="date">2025-03-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pln-team/PLNmodels/blob/master/vignettes/Import_data.Rmd" class="external-link"><code>vignettes/Import_data.Rmd</code></a></small>
      <div class="hidden name"><code>Import_data.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>This vignette documents the data format used in
<strong>PLNmodel</strong> by <code>PLN</code> and its variants. It also
shows how to create an object in the proper format for further analyses
from (i) tabular data, (ii) biom-class objects and (iii) phyloseq-class
objects.</p>
</div>
<div class="section level2">
<h2 id="format-description">Format description<a class="anchor" aria-label="anchor" href="#format-description"></a>
</h2>
<p>We illustrate the format using trichoptera data set, a full
description of which can be found in <a href="Trichoptera.html">the
corresponding vignette</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pln-team.github.io/PLNmodels/">PLNmodels</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">)</span></span></code></pre></div>
<p>The trichoptera data set is a list made of two data frames:
<code>Abundance</code> (hereafter referred to as the <em>counts</em>)
and <code>Covariate</code> (hereafter the <em>covariates</em>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">trichoptera</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 2</span></span>
<span><span class="co">##  $ Abundance:'data.frame':   49 obs. of  17 variables:</span></span>
<span><span class="co">##  $ Covariate:'data.frame':   49 obs. of  7 variables:</span></span></code></pre>
<p>The covariates include, among others, the wind, pressure and
humidity.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Temperature"   "Wind"          "Pressure"      "Humidity"     </span></span>
<span><span class="co">## [5] "Cloudiness"    "Precipitation" "Group"</span></span></code></pre>
<p>In the PLN framework, we model the counts from the covariates, let’s
say wind and pressure, using a Poisson Log-Normal model. Most models in
R use the so-called <em>formula interface</em> and it would thus be
natural to write something like</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PLN.html">PLN</a></span><span class="op">(</span><span class="va">Abundance</span> <span class="op">~</span> <span class="va">Wind</span> <span class="op">+</span> <span class="va">Pressure</span>, data <span class="op">=</span> <span class="va">trichoptera</span><span class="op">)</span></span></code></pre></div>
<p>Unfortunately and unlike many generalized linear models, the response
in PLN is intrinsically <strong>multivariate</strong>: it has 17
dimensions in our example. The left hand side (LHS) must encode a
multivariate response across multiple samples, using a 2D-array (e.g. a
matrix or a data frame).</p>
<p>We must therefore prepare a data structure where
<code>Abundance</code> refers to a count <em>matrix</em> whereas
<code>Wind</code> and <code>Pressure</code> refer to <em>vectors</em>
before feeding it to <code>PLN</code>. That’s the purpose of
<code>prepare_data</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trichoptera2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span>counts     <span class="op">=</span> <span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, </span>
<span>                             covariates <span class="op">=</span> <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">trichoptera2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    49 obs. of  9 variables:</span></span>
<span><span class="co">##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   .. ..$ : chr [1:49] "1" "2" "3" "4" ...</span></span>
<span><span class="co">##   .. ..$ : chr [1:17] "Che" "Hyc" "Hym" "Hys" ...</span></span>
<span><span class="co">##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...</span></span>
<span><span class="co">##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...</span></span>
<span><span class="co">##  $ Pressure     : num  998 1000 997 991 990 ...</span></span>
<span><span class="co">##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...</span></span>
<span><span class="co">##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...</span></span>
<span><span class="co">##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...</span></span>
<span><span class="co">##  $ Group        : Factor w/ 12 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ Offset       : num  29 13 38 192 79 18 8 34 12 4 ...</span></span></code></pre>
<p>If you look carefully, you can notice a few difference between
<code>trichoptera</code> and <code>trichoptera2</code>:</p>
<ul>
<li>the first is a <code>list</code> whereas the second is a
<code>data.frame</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>;</li>
<li>
<code>Abundance</code> is a matrix-column of
<code>trichoptera2</code> that you can extract using the usual functions
<code>[</code> and <code>[[</code> to retrieve the count matrix;</li>
<li>
<code>trichoptera2</code> has an additional <code>Offset</code>
column (more on that later).</li>
</ul>
</div>
<div class="section level2">
<h2 id="computing-offsets">Computing offsets<a class="anchor" aria-label="anchor" href="#computing-offsets"></a>
</h2>
<p>It is common practice when modeling count data to introduce an offset
term to control for different sampling efforts, exposures, baselines,
etc. The <em>proper way</em> to compute sample-specific offsets in still
debated and may vary depending on the field. There are nevertheless a
few popular methods:</p>
<ul>
<li>Total Sum Scaling (TSS), where the offset of a sample is the total
count in that sample</li>
<li>Cumulative Sum Scaling (CSS), introduced in <span class="citation">(<a href="#ref-CSS">Paulson et al. 2013</a>)</span>,
where the offset of a sample if the cumulative sum of counts in that
sample, up to a quantile determined in a data driven way.</li>
<li>Relative Log-Expression (RLE), implemented in <span class="citation">(<a href="#ref-DESeq2">Anders and Huber
2010</a>)</span>, where all samples are used to compute a reference
sample, each sample is compared to the reference sample using log-ratios
and the offset is the median log-ratio.</li>
<li>Geometric Mean of Pairwise Ratio (GMPR), introduced in <span class="citation">(<a href="#ref-GMPR">Chen et al. 2018</a>)</span> where
each sample is compared to each other to compute a median log-ratio and
the offset of a sample is the geometric means of those pairwise
ratios.</li>
<li>Wrench, introduced in <span class="citation">(<a href="#ref-Kumar2018">Kumar et al. 2018</a>)</span> and fully
implemented in the <a href="https://bioconductor.org/packages/release/bioc/html/Wrench.html" class="external-link">Wrench
package</a>, where all samples are used to compute reference proportions
and each sample is compared to the reference using ratios (and
<strong>not log-ratios</strong>) of proportions to compute compositional
correction factors. In that case, the offset is the product of
(geometrically centered) compositional factors and (geometrically
centered) depths.</li>
</ul>
<p>Each of these offset be computed from a counts matrix using the
<code>compute_offset</code> function and changing its
<code>offset</code> argument:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## same as compute_offset(trichoptera$Abundance, offset = "TSS")</span></span>
<span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">##    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 </span></span>
<span><span class="co">##   29   13   38  192   79   18    8   34   12    4    4    3   49   33  600  172 </span></span>
<span><span class="co">##   17   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32 </span></span>
<span><span class="co">##   58   51   56  127   35   13   17    3   27   40   44    8    9 1599 2980   88 </span></span>
<span><span class="co">##   33   34   35   36   37   38   39   40   41   42   43   44   45   46   47   48 </span></span>
<span><span class="co">##  135  327   66   90   63   15   14   20   70   53   95   43   62  149   16   31 </span></span>
<span><span class="co">##   49 </span></span>
<span><span class="co">##   86</span></span></code></pre>
<p>In this particular example, the counts are too sparse and
sophisticated offset methods all fail (numeric output hidden)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"CSS"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in offset_function(counts, ...): Some samples only have 1 positive</span></span>
<span><span class="co">## values. Can't compute quantiles and fall back to TSS normalization</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"RLE"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in offset_function(counts, ...): Because of high sparsity, some samples</span></span>
<span><span class="co">## have null or infinite offset.</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"GMPR"</span><span class="op">)</span></span></code></pre></div>
<p>We can mitigate this problem for the RLE offset by adding
pseudocounts to the counts although doing so has its own drawbacks.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"RLE"</span>, pseudocounts <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         1         2         3         4         5         6         7         8 </span></span>
<span><span class="co">## 0.9186270 0.8349121 0.8570257 0.9186270 0.9186270 0.8349121 0.8192245 0.8570257 </span></span>
<span><span class="co">##         9        10        11        12        13        14        15        16 </span></span>
<span><span class="co">## 0.7322797 0.7322797 0.6321923 0.6321923 0.9240361 0.9186270 1.3788037 0.9240361 </span></span>
<span><span class="co">##        17        18        19        20        21        22        23        24 </span></span>
<span><span class="co">## 0.9186270 0.9240361 0.9240361 1.7140514 0.8908577 0.8570257 0.8349121 0.6321923 </span></span>
<span><span class="co">##        25        26        27        28        29        30        31        32 </span></span>
<span><span class="co">## 0.9240361 0.9240361 0.9186270 0.8570257 0.8349121 2.7721084 3.2934218 0.9584503 </span></span>
<span><span class="co">##        33        34        35        36        37        38        39        40 </span></span>
<span><span class="co">## 1.0406547 0.9584503 0.9584503 0.9186270 0.9584503 0.8908577 0.8349121 0.7322797 </span></span>
<span><span class="co">##        41        42        43        44        45        46        47        48 </span></span>
<span><span class="co">## 0.9240361 0.9240361 0.9584503 0.9584503 1.2643846 1.7140514 0.8233555 0.8908577 </span></span>
<span><span class="co">##        49 </span></span>
<span><span class="co">## 0.9186270</span></span></code></pre>
<p>A better solution consists in using only positive counts to compute
the offsets:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"RLE"</span>, type <span class="op">=</span> <span class="st">"poscounts"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         1         2         3         4         5         6         7         8 </span></span>
<span><span class="co">## 0.5631099 0.9462046 0.8299806 0.9462046 0.6460415 0.5756774 0.6308031 0.4947789 </span></span>
<span><span class="co">##         9        10        11        12        13        14        15        16 </span></span>
<span><span class="co">## 0.7988730 0.3574190 0.2207270 0.1260525 0.8051707 0.7289732 1.5591914 1.1850090 </span></span>
<span><span class="co">##        17        18        19        20        21        22        23        24 </span></span>
<span><span class="co">## 0.6389431 1.0000000 1.4606879 1.8224330 0.8403498 0.3942114 0.4058586 0.1997183 </span></span>
<span><span class="co">##        25        26        27        28        29        30        31        32 </span></span>
<span><span class="co">## 0.6286560 0.5631099 0.8777257 0.3440303 0.2504662 2.7086423 2.5800932 1.4584997 </span></span>
<span><span class="co">##        33        34        35        36        37        38        39        40 </span></span>
<span><span class="co">## 2.6050843 4.5898981 1.2185072 1.2616062 0.8823673 0.6250713 0.3950030 0.5631099 </span></span>
<span><span class="co">##        41        42        43        44        45        46        47        48 </span></span>
<span><span class="co">## 1.4511384 1.0000000 0.9462046 1.0882448 1.0000000 1.0631099 0.4621924 0.7900060 </span></span>
<span><span class="co">##        49 </span></span>
<span><span class="co">## 1.0672361</span></span></code></pre>
<p>Finally, we can use wrench to compute the offsets:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"Wrench"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  0.41269451  0.17385897  0.31925785  0.70682391  0.44749382  0.20676142</span></span>
<span><span class="co">##  [7]  0.10226641  0.21204241  0.09228293  0.03919178  0.03946910  0.02295077</span></span>
<span><span class="co">## [13] 24.18096281  1.54038084  3.95187144  0.80172207 10.52136472  4.55854738</span></span>
<span><span class="co">## [19]  0.46968859 51.88528943  0.32803605  0.14764973  4.13635353  0.03152826</span></span>
<span><span class="co">## [25]  1.63626578  0.53769290  0.38482355  0.62689594  0.08030685 83.87692476</span></span>
<span><span class="co">## [31] 33.40810265  0.75508731  1.05809625  1.18385013  0.58657751  0.43005272</span></span>
<span><span class="co">## [37]  4.79925224  0.18381986  0.16078594  0.18490773  4.25999137  5.46689591</span></span>
<span><span class="co">## [43]  0.60139850  7.56594658  9.89766712 33.73727074  0.74264459 29.13747796</span></span>
<span><span class="co">## [49] 47.12017975</span></span></code></pre>
<p><strong>Note</strong> TSS is the only methods that produces offset on
the same scale as the counts, all others produces offsets that are
(hopefully) <em>proportional</em> to library sizes but on a different
scale. To force the offsets to be on the same scale as the counts for
all methods, you can use the option <code>scale = "count"</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"Wrench"</span>, scale <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            1            2            3            4            5            6 </span></span>
<span><span class="co">##   17.2788462    7.2791915   13.3668057   29.5935647   18.7358369    8.6567635 </span></span>
<span><span class="co">##            7            8            9           10           11           12 </span></span>
<span><span class="co">##    4.2817277    8.8778701    3.8637358    1.6408958    1.6525069    0.9609112 </span></span>
<span><span class="co">##           13           14           15           16           17           18 </span></span>
<span><span class="co">## 1012.4174950   64.4932346  165.4584151   33.5667961  440.5123899  190.8589478 </span></span>
<span><span class="co">##           19           20           21           22           23           24 </span></span>
<span><span class="co">##   19.6650957 2172.3524890   13.7343345    6.1818537  173.1823794    1.3200369 </span></span>
<span><span class="co">##           25           26           27           28           29           30 </span></span>
<span><span class="co">##   68.5077809   22.5123254   16.1119346   26.2471113    3.3623170 3511.7901098 </span></span>
<span><span class="co">##           31           32           33           34           35           36 </span></span>
<span><span class="co">## 1398.7427987   31.6142749   44.3007651   49.5658750   24.5590443   18.0056062 </span></span>
<span><span class="co">##           37           38           39           40           41           42 </span></span>
<span><span class="co">##  200.9368679    7.6962380    6.7318451    7.7417851  178.3588942  228.8900199 </span></span>
<span><span class="co">##           43           44           45           46           47           48 </span></span>
<span><span class="co">##   25.1795750  316.7738500  414.3991879 1412.5245302   31.0933184 1219.9387043 </span></span>
<span><span class="co">##           49 </span></span>
<span><span class="co">## 1972.8451146</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="building-data-frame-using-prepare_data">Building data frame using <code>prepare_data</code><a class="anchor" aria-label="anchor" href="#building-data-frame-using-prepare_data"></a>
</h2>
<p>We’ll already learned that <code>prepare_data</code> can join counts
and covariates into a single data.frame. It can also compute offset
through <code>compute_offset</code> and does so by default with
<code>offset = "TSS"</code>, hence the <code>Offset</code> column in
<code>trichoptera2</code>. You can change the offset method and provide
additional arguments that will passed on to
<code>compute_offset</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, </span>
<span>             <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span>, </span>
<span>             offset <span class="op">=</span> <span class="st">"RLE"</span>, pseudocounts <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    49 obs. of  9 variables:</span></span>
<span><span class="co">##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   .. ..$ : chr [1:49] "1" "2" "3" "4" ...</span></span>
<span><span class="co">##   .. ..$ : chr [1:17] "Che" "Hyc" "Hym" "Hys" ...</span></span>
<span><span class="co">##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...</span></span>
<span><span class="co">##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...</span></span>
<span><span class="co">##  $ Pressure     : num  998 1000 997 991 990 ...</span></span>
<span><span class="co">##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...</span></span>
<span><span class="co">##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...</span></span>
<span><span class="co">##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...</span></span>
<span><span class="co">##  $ Group        : Factor w/ 12 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ Offset       : num  0.919 0.835 0.857 0.919 0.919 ...</span></span></code></pre>
<p>Different communities use different standard for the count data where
samples are either or columns of the counts matrix.
<code>prepare_data</code> uses heuristics to guess the direction of the
counts matrix (or fail informatively doing so) and automatically
transpose it if needed.</p>
<p>Finally, <code>prepare_data</code> enforces sample-consistency
between the counts and the covariates and automatically trims away: -
samples for which only covariates or only counts are available; -
samples with no positive counts</p>
<p>For example, if we remove the first sample from the counts and the
last one from the covariates, we end up with 49 - 2 = 47 samples left,
as expected.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>, <span class="op">]</span>, <span class="co">## remove first sample</span></span>
<span>                  <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">[</span><span class="op">-</span><span class="fl">49</span>,<span class="op">]</span>  <span class="co">## remove last sample</span></span>
<span>                  <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 47</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom">Importing data from biom and phyloseq objects using
<code>prepare_data_from_[phyloseq|biom]</code><a class="anchor" aria-label="anchor" href="#importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom"></a>
</h2>
<p>Community composition data are quite popular in microbial ecology and
usually stored in flat files using the <a href="http://biom-format.org/" class="external-link">biom format</a> and/or imported in R as
phyloseq-class objects <span class="citation">(<a href="#ref-phyloseq">McMurdie 2013</a>)</span> using the Bioconductor <a href="https://joey711.github.io/phyloseq/" class="external-link">phyloseq</a> package.</p>
<!-- We provide helper functions to directly import data from a biom file (or biom-class object) and a phyloseq-class object.  -->
<p>We show here how to import data from a biom file (or biom-class
object) and form a phyloseq-class object.</p>
<div class="section level3">
<h3 id="reading-from-a-biom-file">Reading from a biom file<a class="anchor" aria-label="anchor" href="#reading-from-a-biom-file"></a>
</h3>
<p>Reading from a biom file requires the bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/biomformat.html" class="external-link">biomformat</a>.
This package is <strong>not</strong> a standard dependency of PLNmodels
and needs to be installed separately.</p>
<p>You can easily prepare your data from a biom file using the following
steps:</p>
<ul>
<li>read your biom file with <code>biomformat::read_biom()</code>
</li>
<li>extract the count table with
<code>biomformat::biom_data()</code>
</li>
<li>extract the covariates with
<code>biomformat::sample_metadata()</code> (or build your own)</li>
<li>feed them to <code>prepare_data</code>
</li>
</ul>
<p>as illustrated below:
<!-- Note that the covariates **must** be stored in the biom object as they are automatically extracted, reading a biom without covariates results in an error.  --></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## If biomformat is not installed, uncomment the following lines</span></span>
<span><span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE)) {</span></span>
<span><span class="co">#   install.packages("BiocManager")</span></span>
<span><span class="co"># }</span></span>
<span><span class="co"># BiocManager::install("biomformat")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">biomformat</span><span class="op">)</span></span>
<span><span class="va">biomfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rich_dense_otu_table.biom"</span>, package <span class="op">=</span> <span class="st">"biomformat"</span><span class="op">)</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">biomformat</span><span class="fu">::</span><span class="fu">read_biom</span><span class="op">(</span><span class="va">biomfile</span><span class="op">)</span></span>
<span><span class="co">## extract counts</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu">biomformat</span><span class="fu">::</span><span class="fu">biom_data</span><span class="op">(</span><span class="va">biom</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span><span class="co">## extract covariates (or prepare your own)</span></span>
<span><span class="va">covariates</span> <span class="op">&lt;-</span> <span class="fu">biomformat</span><span class="fu">::</span><span class="fu">sample_metadata</span><span class="op">(</span><span class="va">biom</span><span class="op">)</span></span>
<span><span class="co">## prepare data</span></span>
<span><span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, covariates <span class="op">=</span> <span class="va">covariates</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">my_data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reading-from-a-phyloseq-class-object">Reading from a phyloseq-class object<a class="anchor" aria-label="anchor" href="#reading-from-a-phyloseq-class-object"></a>
</h3>
<p>Likewise, preparing data from a phyloseq-class object requires the
bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html" class="external-link">phyloseq</a>.
This package is <strong>not</strong> a standard dependency of PLNmodels
and needs to be installed separately.</p>
<p>You can easily prepare your data from a phyloseq object using the
following steps:</p>
<ul>
<li>extract the count table with <code>phyloseq::otu_table()</code>
</li>
<li>extract the covariates with <code>phyloseq::sample_data()</code> (or
build your own)</li>
<li>feed them to <code>prepare_data</code>
</li>
</ul>
<p>as illustrated below:</p>
<!-- Note that the covariates **must** be stored in the phyloseq-class object as they are automatically extracted, importing a phyloseq object with no `sample_data` component results in an error.  -->
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## If biomformat is not installed, uncomment the following lines</span></span>
<span><span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE)) {</span></span>
<span><span class="co">#   install.packages("BiocManager")</span></span>
<span><span class="co"># }</span></span>
<span><span class="co"># BiocManager::install("phyloseq")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">phyloseq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"enterotype"</span><span class="op">)</span></span>
<span><span class="co">## extract counts</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu">otu_table</span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span><span class="co">## extract covariates (or prepare your own)</span></span>
<span><span class="va">covariates</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu">sample_data</span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span></span>
<span><span class="co">## prepare data</span></span>
<span><span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, covariates <span class="op">=</span> <span class="va">covariates</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">my_data</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="mathematical-details-about-the-offsets">Mathematical details about the offsets<a class="anchor" aria-label="anchor" href="#mathematical-details-about-the-offsets"></a>
</h2>
<p>We detail here the mathematical background behind the various offsets
and the way they are computed. Note
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝐘</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{Y} = (Y_{ij})</annotation></semantics></math>
the counts matrix where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">Y_{ij}</annotation></semantics></math>
is the count of species
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
in sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
Assume that there are
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
species and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
samples in total. The offset of sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is noted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>O</mi><mi>i</mi></msub><annotation encoding="application/x-tex">O_i</annotation></semantics></math>
and computed in the following way.</p>
<div class="section level3">
<h3 id="total-sum-scaling">Total Sum Scaling<a class="anchor" aria-label="anchor" href="#total-sum-scaling"></a>
</h3>
<p>Offsets are simply the total counts of a sample (frequently called
depths in the metabarcoding literature):
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>O</mi><mi>i</mi></msub><mo>=</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">
O_i = \sum_{j=1}^p Y_{ij}
</annotation></semantics></math></p>
</div>
<div class="section level3">
<h3 id="cumulative-sum-scaling">Cumulative Sum Scaling<a class="anchor" aria-label="anchor" href="#cumulative-sum-scaling"></a>
</h3>
<p>Positive counts are used to compute sample-specific quantiles
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>q</mi><mi>i</mi><mi>l</mi></msubsup><annotation encoding="application/x-tex">q_i^l</annotation></semantics></math>
and cumulative sums
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>s</mi><mi>i</mi><mi>l</mi></msubsup><annotation encoding="application/x-tex">s_i^l</annotation></semantics></math>
defined as
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>q</mi><mi>i</mi><mi>l</mi></msubsup><mo>=</mo><mo>min</mo><mo stretchy="false" form="prefix">{</mo><mi>q</mi><mrow><mspace width="0.333em"></mspace><mtext mathvariant="normal"> such that </mtext><mspace width="0.333em"></mspace></mrow><munder><mo>∑</mo><mi>j</mi></munder><msub><mn>1</mn><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>≤</mo><mi>q</mi></mrow></msub><mo>≥</mo><mi>l</mi><munder><mo>∑</mo><mi>j</mi></munder><msub><mn>1</mn><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>&gt;</mo><mn>0</mn></mrow></msub><mo stretchy="false" form="postfix">}</mo><mspace width="2.0em"></mspace><msubsup><mi>s</mi><mi>i</mi><mi>l</mi></msubsup><mo>=</mo><munder><mo>∑</mo><mrow><mi>j</mi><mo>:</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>≤</mo><msubsup><mi>q</mi><mi>i</mi><mi>l</mi></msubsup></mrow></munder><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">
q_i^l = \min \{q \text{ such that } \sum_j 1_{Y_{ij} \leq q} \geq l \sum_j 1_{Y_{ij} &gt; 0} \} \qquad s_i^l = \sum_{j: Y_{ij} \leq q_i^l} Y_{ij}
</annotation></semantics></math> The sample-specific quantiles are then
used to compute reference quantiles defined as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>q</mi><mi>l</mi></msup><mo>=</mo><mtext mathvariant="normal">median</mtext><mo stretchy="false" form="prefix">{</mo><msubsup><mi>q</mi><mi>l</mi><mi>i</mi></msubsup><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">q^l = \text{median} \{q^i_l\}</annotation></semantics></math>
and median average deviation around the quantile
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>q</mi><mi>l</mi></msup><annotation encoding="application/x-tex">q^l</annotation></semantics></math>
as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>d</mi><mi>l</mi></msup><mo>=</mo><mtext mathvariant="normal">median</mtext><mrow><mo stretchy="true" form="prefix">|</mo><msubsup><mi>q</mi><mi>i</mi><mi>l</mi></msubsup><mo>−</mo><msup><mi>q</mi><mi>l</mi></msup><mo stretchy="true" form="postfix">|</mo></mrow></mrow><annotation encoding="application/x-tex">d^l = \text{median} |q_i^l - q^l|</annotation></semantics></math>.
The method then searches for the smallest quantile
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>l</mi><annotation encoding="application/x-tex">l</annotation></semantics></math>
for which it detects instability, defined as large relative increase in
the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>d</mi><mi>l</mi></msup><annotation encoding="application/x-tex">d^l</annotation></semantics></math>.
Formally,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>l</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{l}</annotation></semantics></math>
is the smallest
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>l</mi><annotation encoding="application/x-tex">l</annotation></semantics></math>
satisfying
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><msup><mi>d</mi><mrow><mi>l</mi><mo>+</mo><mn>1</mn></mrow></msup><mo>−</mo><msup><mi>d</mi><mi>l</mi></msup></mrow><msup><mi>d</mi><mi>l</mi></msup></mfrac><mo>≥</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">\frac{d^{l+1} - d^l}{d^l} \geq 0.1</annotation></semantics></math>.
The scaling sample-specific offset are then chosen as:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>O</mi><mi>i</mi></msub><mo>=</mo><msubsup><mi>s</mi><mi>i</mi><mover><mi>l</mi><mo accent="true">̂</mo></mover></msubsup><mi>/</mi><msub><mtext mathvariant="normal">median</mtext><mi>i</mi></msub><mo stretchy="false" form="prefix">{</mo><msubsup><mi>s</mi><mi>i</mi><mover><mi>l</mi><mo accent="true">̂</mo></mover></msubsup><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">
O_i = s_i^{\hat{l}} / \text{median}_i \{ s_i^{\hat{l}} \}
</annotation></semantics></math> Dividing by the median of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>s</mi><mi>i</mi><mover><mi>l</mi><mo accent="true">̂</mo></mover></msubsup><annotation encoding="application/x-tex">s_i^{\hat{l}}</annotation></semantics></math>
ensures that offsets are centered around
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math>
and compare sizes differences with respect to the reference sample. Note
also that the reference quantiles
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>q</mi><mi>l</mi></msup><annotation encoding="application/x-tex">q^l</annotation></semantics></math>
can be computed using either the median (default, as in the original
<span class="citation">Paulson et al. (<a href="#ref-CSS">2013</a>)</span> paper) or the mean, by specifying
<code>reference = mean</code>, as implemented in
<code>metagenomeseq</code>.</p>
</div>
<div class="section level3">
<h3 id="relative-log-expression">Relative Log Expression<a class="anchor" aria-label="anchor" href="#relative-log-expression"></a>
</h3>
<p>A reference sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>q</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mi>j</mi></msub><annotation encoding="application/x-tex">(q_j)_j</annotation></semantics></math>
is first built by computing the geometric means of each species count:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>q</mi><mi>j</mi></msub><mo>=</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><mn>1</mn><mi>n</mi></mfrac><munder><mo>∑</mo><mi>i</mi></munder><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
q_j = \exp \left( \frac{1}{n} \sum_{i} \log(Y_{ij})\right)
</annotation></semantics></math> Each sample is then compared to the
reference sample to compute one ratio per species and the final offset
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>O</mi><mi>i</mi></msub><annotation encoding="application/x-tex">O_i</annotation></semantics></math>
is the median of those ratios:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>O</mi><mi>i</mi></msub><mo>=</mo><msub><mtext mathvariant="normal">median</mtext><mi>j</mi></msub><mfrac><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>q</mi><mi>j</mi></msub></mfrac></mrow><annotation encoding="application/x-tex">
O_i = \text{median}_j \frac{Y_{ij}}{q_j}
</annotation></semantics></math> The method fails when no species is
shared across all sample (as all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>q</mi><mi>j</mi></msub><annotation encoding="application/x-tex">q_j</annotation></semantics></math>
are then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>)
or when a sample shares less than 50% of species with the reference (in
which case the median of the ratios may be null or infinite). The
problem can be alleviated by adding pseudocounts to the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>c</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">c_{ij}</annotation></semantics></math>
with <code>pseudocounts = 1</code> or using positive counts in the
computations (<code>type = "poscounts"</code>)</p>
</div>
<div class="section level3">
<h3 id="geometric-mean-of-pairwise-ratio">Geometric Mean of Pairwise Ratio<a class="anchor" aria-label="anchor" href="#geometric-mean-of-pairwise-ratio"></a>
</h3>
<p>This method is similar to RLE but does create a reference sample.
Instead, each sample is compared to each other to compute a median ratio
(similar to RLE)
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>r</mi><mrow><mi>i</mi><mi>i</mi><mi>′</mi></mrow></msub><mo>=</mo><msub><mtext mathvariant="normal">median</mtext><mrow><mi>j</mi><mo>:</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mi>.</mi><msub><mi>Y</mi><mrow><mi>i</mi><mi>′</mi><mi>j</mi></mrow></msub><mo>&gt;</mo><mn>0</mn></mrow></msub><mfrac><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>Y</mi><mrow><mi>i</mi><mi>′</mi><mi>j</mi></mrow></msub></mfrac></mrow><annotation encoding="application/x-tex">
r_{ii'} = {\text{median}}_{j: Y_{ij}.Y_{i'j} &gt; 0} \frac{Y_{ij}}{Y_{i'j}}
</annotation></semantics></math> The offset is then taken as the median
of all the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>i</mi><mi>′</mi></mrow></msub><annotation encoding="application/x-tex">r_{ii'}</annotation></semantics></math>:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>O</mi><mi>i</mi></msub><mo>=</mo><msub><mtext mathvariant="normal">median</mtext><mrow><mi>i</mi><mi>′</mi><mi>!</mi><mo>=</mo><mi>i</mi></mrow></msub><msub><mi>r</mi><mrow><mi>i</mi><mi>i</mi><mi>′</mi></mrow></msub></mrow><annotation encoding="application/x-tex">
O_i = \text{median}_{i' != i} r_{ii'}
</annotation></semantics></math> The method fails when there is only one
sample in the data set or when a sample shares no species with any
other.</p>
</div>
<div class="section level3">
<h3 id="wrench-normalisation">Wrench normalisation<a class="anchor" aria-label="anchor" href="#wrench-normalisation"></a>
</h3>
<p>This method is fully detailed in <span class="citation">Kumar et al.
(<a href="#ref-Kumar2018">2018</a>)</span> and we only provide a
barebone implementation corresponding to the defaults parameters of
<code>Wrench::wrench()</code>. Assume that samples belong to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
discrete groups and note
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>g</mi><mi>i</mi></msub><annotation encoding="application/x-tex">g_i</annotation></semantics></math>
the group of sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
Wrench is based on the following (simplified) log-normal model for
counts:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>∼</mo><msub><mi>π</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>δ</mi><mn>0</mn></msub><mo>+</mo><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><msub><mi>π</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>log</mo><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>μ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>,</mo><msubsup><mi>σ</mi><mi>j</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
Y_{ij} \sim \pi_{ij} \delta_0 + (1 - \pi_{ij})\log\mathcal{N}(\mu_{ij}, \sigma^2_j)
</annotation></semantics></math> where the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">Y_{ij}</annotation></semantics></math>
are independent and the mean
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{ij}</annotation></semantics></math>
is decomposed as:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>μ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><munder><munder><mrow><mo>log</mo><msub><mi>p</mi><mrow><mn>0</mn><mi>j</mi></mrow></msub></mrow><mo accent="true">⏟</mo></munder><mtext mathvariant="normal">log-ref. prop.</mtext></munder><mo>+</mo><munder><munder><mrow><mo>log</mo><msub><mi>d</mi><mi>i</mi></msub></mrow><mo accent="true">⏟</mo></munder><mtext mathvariant="normal">log-depth</mtext></munder><mo>+</mo><munder><munder><mrow><mo>log</mo><msub><mi>ζ</mi><mrow><mn>0</mn><msub><mi>g</mi><mi>i</mi></msub></mrow></msub></mrow><mo accent="true">⏟</mo></munder><mrow><mrow><mtext mathvariant="normal">log effect of group </mtext><mspace width="0.333em"></mspace></mrow><msub><mi>g</mi><mi>i</mi></msub></mrow></munder><mo>+</mo><munder><munder><msub><mi>a</mi><mi>i</mi></msub><mo accent="true">⏟</mo></munder><mtext mathvariant="normal">(f|m)ixed effect</mtext></munder><mo>+</mo><munder><munder><msub><mi>b</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo accent="true">⏟</mo></munder><mtext mathvariant="normal">mixed effects</mtext></munder></mrow><annotation encoding="application/x-tex">
\mu_{ij} = \underbrace{\log{p_{0j}}}_{\text{log-ref. prop.}} +  \underbrace{\log{d_i}}_{\text{log-depth}} + \underbrace{\log{\zeta_{0g_i}}}_{\text{log effect of group } g_i} + \underbrace{a_{i}}_{\text{(f|m)ixed effect}} + \underbrace{b_{ij}}_{\text{mixed effects}}
</annotation></semantics></math> where the random effects are
independents centered gaussian and the depths is the total sum of
counts:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>d</mi><mi>i</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><msub><mi>c</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>b</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>∼</mo><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><msubsup><mi>η</mi><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align*}
d_i &amp; = \sum_{j=1}^p c_{ij} \\
b_{ij} &amp; \sim \mathcal{N}(0, \eta^2_{g_i}) \\
\end{align*}
</annotation></semantics></math></p>
<p>The <strong>net</strong> log fold change
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>θ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\theta_{ij}</annotation></semantics></math>
of the <strong>proportion ratio</strong>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><msub><mi>c</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mi>/</mi><msub><mi>d</mi><mi>i</mi></msub><msub><mi>p</mi><mrow><mn>0</mn><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">r_{ij} = c_{ij} / d_i p_{0j}</annotation></semantics></math>
of species
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
relative to the reference is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>θ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mover><mo>=</mo><mi>Δ</mi></mover><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="false" form="prefix">|</mo><msub><mi>a</mi><mi>i</mi></msub><mo>,</mo><msub><mi>b</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mo>log</mo><msub><mi>ζ</mi><mrow><mn>0</mn><msub><mi>g</mi><mi>i</mi></msub></mrow></msub><mo>+</mo><msub><mi>a</mi><mi>i</mi></msub><mo>+</mo><msub><mi>b</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\log(\theta_{ij}) \overset{\Delta}{=} \mathbb{E}[\log(r_{ij}) | a_i, b_{ij}] = \log{\zeta_{0g_i}} + a_i + b_{ij}</annotation></semantics></math>.
We can decompose it as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>θ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><msubsup><mi>Λ</mi><mi>i</mi><mrow><mo>−</mo><mn>1</mn></mrow></msubsup><msub><mi>v</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\theta_{ij} = \Lambda_i^{-1} v_{ij}</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>Λ</mi><mi>i</mi><mrow><mo>−</mo><mn>1</mn></mrow></msubsup><annotation encoding="application/x-tex">\Lambda_i^{-1}</annotation></semantics></math>
is the <em>compositional correction factor</em> and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>v</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">v_{ij}</annotation></semantics></math>
is the fold change of <strong>true abundances</strong>.</p>
<p>With the above notations, the net fold change compounds both the fold
change of true abundances and the compositional correction factors. With
the assumption that the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>b</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">b_{ij}</annotation></semantics></math>
are centered,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mover><mi>Λ</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\log(\hat{\Lambda}_i)</annotation></semantics></math>
can be estimated through a robust average of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mover><mi>θ</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\hat{\theta}_{ij}</annotation></semantics></math>,
which can themselves be computed from the log-ratio of proportions.</p>
<p>We detail here how the different parameters and/or effects are
estimated.</p>
<ul>
<li>The reference proportions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mrow><mn>0</mn><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">p_{0j}</annotation></semantics></math>
are constructed as averages of the sample proportions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">p_{ij}</annotation></semantics></math>
and the ratio are derived from both quantities
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mspace width="2.0em"></mspace><msub><mi>p</mi><mrow><mn>0</mn><mi>j</mi></mrow></msub><mo>=</mo><mfrac><mn>1</mn><mi>n</mi></mfrac><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>p</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="2.0em"></mspace><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mi>p</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>p</mi><mrow><mn>0</mn><mi>j</mi></mrow></msub></mfrac></mrow><annotation encoding="application/x-tex">
p_{ij} = \frac{Y_{ij}}{\sum_{j=1}^p Y_{ij}} \qquad p_{0j} = \frac{1}{n} \sum_{i=1}^n p_{ij} \qquad r_{ij} = \frac{p_{ij}}{p_{0j}}
</annotation></semantics></math>
</li>
<li>The probabilities of absence
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{ij}</annotation></semantics></math>
are estimated by fitting the following Bernoulli models:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mn>1</mn><mrow><mo stretchy="false" form="prefix">{</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mn>0</mn><mo stretchy="false" form="postfix">}</mo></mrow></msub><mo>∼</mo><mi>ℬ</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>π</mi><mi>j</mi><msub><mi>d</mi><mi>i</mi></msub></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
1_{\{Y_{ij} = 0\}} \sim \mathcal{B}(\pi_{j}^{d_i})
</annotation></semantics></math> and setting
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><msup><mover><mi>π</mi><mo accent="true">̂</mo></mover><msub><mi>d</mi><mi>i</mi></msub></msup></mrow><annotation encoding="application/x-tex">\hat{\pi}_{ij} = \hat{\pi}^{d_i}</annotation></semantics></math>
</li>
<li>The species variances
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>j</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_j</annotation></semantics></math>
are estimated by fitting the following linear model (with no
zero-inflation component)
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>∼</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>d</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><msub><mi>μ</mi><msub><mi>g</mi><mi>i</mi></msub></msub><mo>+</mo><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><msubsup><mi>σ</mi><mi>j</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
\log Y_{ij} \sim \log(d_i) + \mu_{g_i} + \mathcal{N}(0, \sigma^2_j)
</annotation></semantics></math> Note that in the original
<code>Wrench::wrench()</code>, the log depth
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>d</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\log(d_i)</annotation></semantics></math>
is used as predictor but I believe it makes more sense to use it an
offset.</li>
<li>set the group proportions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">p_{gj}</annotation></semantics></math>
and group ratios
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">r_{gj}</annotation></semantics></math>
to:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mrow><mi>i</mi><mo>:</mo><msub><mi>g</mi><mi>i</mi></msub><mo>=</mo><mi>g</mi></mrow></munder><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><mrow><munder><mo>∑</mo><mrow><mi>j</mi><mo>,</mo><mi>i</mi><mo>:</mo><msub><mi>g</mi><mi>i</mi></msub><mo>=</mo><mi>g</mi></mrow></munder><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mspace width="2.0em"></mspace><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msub><mi>p</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><msub><mi>p</mi><mrow><mn>0</mn><mi>j</mi></mrow></msub></mfrac></mrow><annotation encoding="application/x-tex">
p_{gj} = \frac{\sum_{i : g_i = g} Y_{ij}}{\sum_{j, i : g_i = g} Y_{ij}} \qquad r_{gj} = \frac{p_{gj}}{p_{0j}}
</annotation></semantics></math>
</li>
<li>Estimate the location and dispersion parameters as:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>ζ</mi><mo accent="true">̂</mo></mover><mrow><mn>0</mn><mi>g</mi></mrow></msub><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub></mrow><mi>p</mi></mfrac><mspace width="2.0em"></mspace><mo>log</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>.</mi></mrow></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mrow><mi>j</mi><mo>:</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>≠</mo><mn>1</mn></mrow></munder><mo>log</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub></mrow><mrow><munder><mo>∑</mo><mrow><mi>j</mi><mo>:</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>≠</mo><mn>0</mn></mrow></munder><mn>1</mn></mrow></mfrac><mspace width="2.0em"></mspace><msubsup><mover><mi>η</mi><mo accent="true">̂</mo></mover><mi>g</mi><mn>2</mn></msubsup><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mrow><mi>j</mi><mo>:</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>≠</mo><mn>1</mn></mrow></munder><msup><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>−</mo><mo>log</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>.</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><mrow><munder><mo>∑</mo><mrow><mi>j</mi><mo>:</mo><msub><mi>r</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>≠</mo><mn>0</mn></mrow></munder><mn>1</mn></mrow></mfrac></mrow><annotation encoding="application/x-tex">
\hat{\zeta}_{0g} = \frac{\sum_{j=1}^p r_{gj}}{p} 
\qquad
\log{r_{g.}} = \frac{\sum_{j: r_{gj} \neq 1} \log{r_{gj}}}{\sum_{j: r_{gj} \neq 0} 1}   
\qquad
\hat{\eta}_{g}^2 = \frac{\sum_{j: r_{gj} \neq 1} (\log{r_{gj}} - \log{r_{g.}})^2}{\sum_{j: r_{gj} \neq 0} 1}
</annotation></semantics></math>
</li>
<li>Estimate the mixed effects as shrunken (and scaled) averages of the
ratios
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>a</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo>=</mo><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><mfrac><mn>1</mn><mrow><msubsup><mover><mi>η</mi><mo accent="true">̂</mo></mover><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>j</mi><mn>2</mn></msubsup></mrow></mfrac><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>−</mo><mo>log</mo><msub><mover><mi>ζ</mi><mo accent="true">̂</mo></mover><mrow><mn>0</mn><msub><mi>g</mi><mi>i</mi></msub></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><mfrac><mn>1</mn><mrow><msubsup><mover><mi>η</mi><mo accent="true">̂</mo></mover><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>j</mi><mn>2</mn></msubsup></mrow></mfrac></mrow></mfrac><mspace width="2.0em"></mspace><msub><mover><mi>b</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mfrac><msubsup><mover><mi>η</mi><mo accent="true">̂</mo></mover><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><mrow><msubsup><mover><mi>η</mi><mo accent="true">̂</mo></mover><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>j</mi><mn>2</mn></msubsup></mrow></mfrac><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>−</mo><mo>log</mo><msub><mover><mi>ζ</mi><mo accent="true">̂</mo></mover><mrow><mn>0</mn><msub><mi>g</mi><mi>i</mi></msub></mrow></msub><mo>−</mo><msub><mover><mi>a</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
\hat{a}_i = \frac{\sum_{j = 1}^p \frac{1}{\hat{\eta}^2_{g_i} + \hat{\sigma}^2_j} (\log{r_{ij}} - \log{\hat{\zeta}_{0g_i}})}{\sum_{j = 1}^p \frac{1}{\hat{\eta}^2_{g_i} + \hat{\sigma}^2_j}}
\qquad 
\hat{b}_{ij} = \frac{\hat{\eta}^2_{g_i}}{\hat{\eta}^2_{g_i} + \hat{\sigma}^2_j} \left( \log{r_{ij}} - \log\hat{\zeta}_{0g_i} - \hat{a}_i\right)
</annotation></semantics></math>
</li>
<li>Estimate the regularized ratios as:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>θ</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><msub><mover><mi>ζ</mi><mo accent="true">̂</mo></mover><mrow><mn>0</mn><msub><mi>g</mi><mi>i</mi></msub></mrow></msub><mo>+</mo><msub><mover><mi>a</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo>+</mo><msub><mover><mi>b</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
\hat{\theta}_{ij} = \exp\left( \log\hat{\zeta}_{0g_i} + \hat{a}_i + \hat{b}_{ij} \right)
</annotation></semantics></math>
</li>
<li>Estimate the compositional correction factors as (weighted) means of
the regularized (and possibly corrected) ratios:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>Λ</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo>=</mo><mrow><mo stretchy="true" form="prefix">{</mo><mtable><mtr><mtd columnalign="left" style="text-align: left"><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><msub><mover><mi>θ</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mi minsize="2.4" maxsize="2.4">/</mi><mi>p</mi></mtd><mtd columnalign="left" style="text-align: left"><mrow><mspace width="0.333em"></mspace><mtext mathvariant="normal"> if type = "simple"</mtext></mrow></mtd></mtr><mtr><mtd columnalign="left" style="text-align: left"><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><msub><mover><mi>θ</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><msup><mi>e</mi><mrow><mo>−</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>j</mi><mn>2</mn></msubsup><mi>/</mi><mn>2</mn></mrow></msup><mi>/</mi><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mi minsize="2.4" maxsize="2.4">/</mi><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>p</mi></munderover><mn>1</mn><mi>/</mi><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mtd><mtd columnalign="left" style="text-align: left"><mrow><mspace width="0.333em"></mspace><mtext mathvariant="normal"> if type = "wrench"</mtext></mrow></mtd></mtr><mtr><mtd columnalign="left" style="text-align: left"></mtd></mtr></mtable></mrow></mrow><annotation encoding="application/x-tex">
\hat{\Lambda}_i = 
\begin{cases}
\sum_{j = 1}^p \hat{\theta}_{ij} \bigg/ p&amp; \text{ if type = "simple"} \\
\sum_{j = 1}^p \hat{\theta}_{ij} e^{-\hat{\sigma}_j^2 / 2} /  w_{ij} \bigg/ \sum_{j=1}^p 1/w_{ij}  &amp; \text{ if type = "wrench"} \\
\end{cases}
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>+</mo><msup><mi>e</mi><mrow><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>j</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>η</mi><mo accent="true">̂</mo></mover><mi>i</mi><mn>2</mn></msubsup></mrow></msup><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">w_{ij} = (1 - \hat{\pi}_{ij})(\hat{\pi}_{ij} + e^{\hat{\sigma}_j^2 + \hat{\eta}_i^2} - 1)</annotation></semantics></math>.
The correction term
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>e</mi><mrow><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>j</mi><mn>2</mn></msubsup><mi>/</mi><mn>2</mn></mrow></msup><annotation encoding="application/x-tex">e^{\hat{\sigma}_j^2 / 2}</annotation></semantics></math>
arises from the relation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>&gt;</mo><mn>0</mn><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><msub><mi>θ</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msup><mi>e</mi><mrow><msubsup><mi>σ</mi><mi>j</mi><mn>2</mn></msubsup><mi>/</mi><mn>2</mn></mrow></msup></mrow><annotation encoding="application/x-tex">\mathbb{E}[r_{ij} | r_{ij} &gt; 0] = \theta_{ij} e^{\sigma_j^2/2}</annotation></semantics></math>
and the weight
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">w_{ij}</annotation></semantics></math>
are marginal variances:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝕍</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\mathbb{V}[r_{ij}] = w_{ij}</annotation></semantics></math>.</li>
</ul>
<p>The offsets are then the product of compositional correction factors
and depths:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>O</mi><mi>i</mi></msub><mo>=</mo><mfrac><msub><mover><mi>Λ</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><msup><mrow><mo stretchy="true" form="prefix">(</mo><munderover><mo>∏</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mover><mi>Λ</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mi>/</mi><mi>n</mi></mrow></msup></mfrac><mo>×</mo><mfrac><msub><mi>d</mi><mi>i</mi></msub><msup><mrow><mo stretchy="true" form="prefix">(</mo><munderover><mo>∏</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>d</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mi>/</mi><mi>n</mi></mrow></msup></mfrac></mrow><annotation encoding="application/x-tex">
O_i = \frac{\hat{\Lambda}_i}{(\prod_{i = 1}^n \hat{\Lambda}_i)^{1/n}} \times \frac{d_i}{(\prod_{i = 1}^n d_i)^{1/n}}
</annotation></semantics></math></p>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-DESeq2" class="csl-entry">
Anders, Simon, and Wolfgang Huber. 2010. <span>“Differential Expression
Analysis for Sequence Count Data.”</span> <em>Genome Biology</em> 11
(10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106" class="external-link">https://doi.org/10.1186/gb-2010-11-10-r106</a>.
</div>
<div id="ref-GMPR" class="csl-entry">
Chen, Li, James Reeve, Lujun Zhang, Shengbing Huang, Xuefeng Wang, and
Jun Chen. 2018. <span>“GMPR: A Robust Normalization Method for
Zero-Inflated Count Data with Application to Microbiome Sequencing
Data.”</span> <em>PeerJ</em> 6 (April): e4600. <a href="https://doi.org/10.7717/peerj.4600" class="external-link">https://doi.org/10.7717/peerj.4600</a>.
</div>
<div id="ref-Kumar2018" class="csl-entry">
Kumar, M. Senthil, Eric V. Slud, Kwame Okrah, Stephanie C. Hicks,
Sridhar Hannenhalli, and Héctor Corrada Bravo. 2018. <span>“Analysis and
Correction of Compositional Bias in Sparse Sequencing Count
Data.”</span> <em><span>BMC</span> Genomics</em> 19 (1). <a href="https://doi.org/10.1186/s12864-018-5160-5" class="external-link">https://doi.org/10.1186/s12864-018-5160-5</a>.
</div>
<div id="ref-phyloseq" class="csl-entry">
McMurdie, Paul J. AND Holmes. 2013. <span>“Phyloseq: An r Package for
Reproducible Interactive Analysis and Graphics of Microbiome Census
Data.”</span> <em>PLoS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217" class="external-link">https://doi.org/10.1371/journal.pone.0061217</a>.
</div>
<div id="ref-CSS" class="csl-entry">
Paulson, Joseph N, O. Colin Stine, Héctor Corrada Bravo, and Mihai Pop.
2013. <span>“Differential Abundance Analysis for Microbial Marker-Gene
Surveys.”</span> <em>Nat Methods</em> 10 (September): 1200–1202. <a href="https://doi.org/10.1038/nmeth.2658" class="external-link">https://doi.org/10.1038/nmeth.2658</a>.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>although a <code>data.frame</code> is technically a
<code>list</code><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Julien Chiquet, Mahendra Mariadassou, Stéphane Robin, François Gindraud.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
