<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data importation in PLNmodels • PLNmodels</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data importation in PLNmodels">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PLNmodels</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.2.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a>
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Trichoptera.html">The trichoptera data set</a>
    </li>
    <li>
      <a href="../articles/Import_data.html">Data importation in PLNmodels</a>
    </li>
    <li>
      <a href="../articles/PLN.html">PLN: Multivariate Poisson regression</a>
    </li>
    <li>
      <a href="../articles/PLNPCA.html">PLNPCA: Dimension Reduction</a>
    </li>
    <li>
      <a href="../articles/PLNnetwork.html">PLNnetwork: sparse structure inference</a>
    </li>
    <li>
      <a href="../articles/PLNLDA.html">PLNLDA: Discriminant Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jchiquet/PLNmodels">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Data importation in PLNmodels</h1>
                        <h4 class="author">PLN team</h4>
            
            <h4 class="date">2019-07-03</h4>
      
      
      <div class="hidden name"><code>Import_data.Rmd</code></div>

    </div>

    
    
<div id="preliminaries" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminaries" class="anchor"></a>Preliminaries</h2>
<p>This vignette documents the data format used in <strong>PLNmodel</strong> by <code>PLN</code> and its variants. It also shows how to create an object in the proper format for further analyses from (i) tabular data, (ii) biom-class objects and (iii) phyloseq-class objects.</p>
</div>
<div id="format-description" class="section level2">
<h2 class="hasAnchor">
<a href="#format-description" class="anchor"></a>Format description</h2>
<p>We illustrate the format using trichoptera data set, a full description of which can be found in <a href="Trichoptera.html">the corresponding vignette</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(PLNmodels)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(trichoptera)</a></code></pre></div>
<p>The trichoptera data set is a list made of two data frames: <code>Abundance</code> (hereafter referred to as the <em>counts</em>) and <code>Covariate</code> (hereafter the <em>covariates</em>).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(trichoptera, <span class="dt">max.level =</span> <span class="dv">1</span>)</a></code></pre></div>
<pre><code>## List of 2
##  $ Abundance:'data.frame':   49 obs. of  17 variables:
##  $ Covariate:'data.frame':   49 obs. of  7 variables:</code></pre>
<p>The covariates include, among others, the wind, pressure and humidity.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(trichoptera<span class="op">$</span>Covariate)</a></code></pre></div>
<pre><code>## [1] "Temperature"   "Wind"          "Pressure"      "Humidity"     
## [5] "Cloudiness"    "Precipitation" "Group"</code></pre>
<p>In the PLN framework, we model the counts from the covariates, let’s say wind and pressure, using a Poisson Log-Normal model. Most models in R use the so-called <em>formula interface</em> and it would thus be naturel to write something like</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="../reference/PLN.html">PLN</a></span>(Abundance <span class="op">~</span><span class="st"> </span>Wind <span class="op">+</span><span class="st"> </span>Pressure, <span class="dt">data =</span> trichoptera)</a></code></pre></div>
<p>Unfortunately and unlike many generalized linear models, the response in PLN is intrinsically <strong>multivariate</strong>: it has 17 dimensions in our example. The left hand side (LHS) must encode a multivariate response across multiple samples, using a 2D-array (e.g. a matrix or a data frame).</p>
<p>We must therefore prepare a data structure where <code>Abundance</code> refers to a count <em>matrix</em> whereas <code>Wind</code> and <code>Pressure</code> refer to <em>vectors</em> before feeding it to <code>PLN</code>. That’s the purpose of <code>prepare_data</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">trichoptera2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepare_data.html">prepare_data</a></span>(<span class="dt">counts     =</span> trichoptera<span class="op">$</span>Abundance, </a>
<a class="sourceLine" id="cb7-2" data-line-number="2">                             <span class="dt">covariates =</span> trichoptera<span class="op">$</span>Covariate)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(trichoptera2)</a></code></pre></div>
<pre><code>## 'data.frame':    49 obs. of  9 variables:
##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr  "1" "2" "3" "4" ...
##   .. ..$ : chr  "Che" "Hyc" "Hym" "Hys" ...
##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...
##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...
##  $ Pressure     : num  998 1000 997 991 990 ...
##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...
##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...
##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...
##  $ Group        : Factor w/ 12 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Offset       : num  29 13 38 192 79 18 8 34 12 4 ...</code></pre>
<p>If you look carefully, you can notice a few difference between <code>trichoptera</code> and <code>trichoptera2</code>:</p>
<ul>
<li>the first is a <code>list</code> whereas the second is a <code>data.frame</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>;</li>
<li>
<code>Abundance</code> is a matrix-column of <code>trichoptera2</code> that you can extract using the usual functions <code>[</code> and <code>[[</code> to retrieve the count matrix;</li>
<li>
<code>trichoptera2</code> has an additional <code>Offset</code> column (more on that later).</li>
</ul>
</div>
<div id="computing-offsets" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-offsets" class="anchor"></a>Computing offsets</h2>
<p>It is common practice when modeling count data to introduce an offset term to control for different sampling efforts, exposures, baselines, etc. The <em>proper way</em> to compute sample-specific offsets in still debated and may vary depending on the field. There are nevertheless a few popular methods:</p>
<ul>
<li>Total Sum Scaling (TSS), where the offset of a sample is the total count in that sample</li>
<li>Cumulative Sum Scaling (CSS), introduced in <span class="citation">(Paulson et al. <a href="#ref-CSS">2013</a>)</span>, where the offset of a sample if the cumulative sum of counts in that sample, up to a quantile determined in a data driven way.</li>
<li>Relative Log-Expression (RLE), implemented in <span class="citation">(Anders and Huber <a href="#ref-DESeq2">2010</a>)</span>, where all samples are used to compute a reference sample, each sample is compared to the reference sample using log-ratios and the offset is the median log-ratio.</li>
<li>Geometric Mean of Pairwise Ratio (GMPR), introduced in <span class="citation">(Chen et al. <a href="#ref-GMPR">2018</a>)</span> where each sample is compared to each other to compute a median log-ratio and the offset of a sample is the geometric means of those pairwise ratios.</li>
</ul>
<p>Each of these offset be computed from a counts matrix using the <code>compute_offset</code> function and changing its <code>offset</code> argument:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">## same as compute_offset(trichoptera$Abundance, offset = "TSS")</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="../reference/compute_offset.html">compute_offset</a></span>(trichoptera<span class="op">$</span>Abundance) </a></code></pre></div>
<pre><code>##    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15 
##   29   13   38  192   79   18    8   34   12    4    4    3   49   33  600 
##   16   17   18   19   20   21   22   23   24   25   26   27   28   29   30 
##  172   58   51   56  127   35   13   17    3   27   40   44    8    9 1599 
##   31   32   33   34   35   36   37   38   39   40   41   42   43   44   45 
## 2980   88  135  327   66   90   63   15   14   20   70   53   95   43   62 
##   46   47   48   49 
##  149   16   31   86</code></pre>
<p>In this particular example, the counts are too sparse and sophisticated offset methods all fail (numeric output hidden)</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="../reference/compute_offset.html">compute_offset</a></span>(trichoptera<span class="op">$</span>Abundance, <span class="st">"CSS"</span>)</a></code></pre></div>
<pre><code>## Warning in offset_function(counts, ...): Some samples only have 1 positive
## values. Can't compute quantiles and fall back to TSS normalization</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="../reference/compute_offset.html">compute_offset</a></span>(trichoptera<span class="op">$</span>Abundance, <span class="st">"RLE"</span>)</a></code></pre></div>
<pre><code>## Warning in offset_function(counts, ...): Because of high sparsity, some
## samples have null or infinite offset.</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../reference/compute_offset.html">compute_offset</a></span>(trichoptera<span class="op">$</span>Abundance, <span class="st">"GMPR"</span>)</a></code></pre></div>
<p>We can mitigate this problem for the RLE offset by adding pseudocounts to the counts although doing so has its own drawbacks.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="../reference/compute_offset.html">compute_offset</a></span>(trichoptera<span class="op">$</span>Abundance, <span class="st">"RLE"</span>, <span class="dt">pseudocounts =</span> <span class="dv">1</span>)</a></code></pre></div>
<pre><code>##         1         2         3         4         5         6         7 
## 0.9186270 0.8349121 0.8570257 0.9186270 0.9186270 0.8349121 0.8192245 
##         8         9        10        11        12        13        14 
## 0.8570257 0.7322797 0.7322797 0.6321923 0.6321923 0.9240361 0.9186270 
##        15        16        17        18        19        20        21 
## 1.3788037 0.9240361 0.9186270 0.9240361 0.9240361 1.7140514 0.8908577 
##        22        23        24        25        26        27        28 
## 0.8570257 0.8349121 0.6321923 0.9240361 0.9240361 0.9186270 0.8570257 
##        29        30        31        32        33        34        35 
## 0.8349121 2.7721084 3.2934218 0.9584503 1.0406547 0.9584503 0.9584503 
##        36        37        38        39        40        41        42 
## 0.9186270 0.9584503 0.8908577 0.8349121 0.7322797 0.9240361 0.9240361 
##        43        44        45        46        47        48        49 
## 0.9584503 0.9584503 1.2643846 1.7140514 0.8233555 0.8908577 0.9186270</code></pre>
</div>
<div id="building-data-frame-using-prepare_data" class="section level2">
<h2 class="hasAnchor">
<a href="#building-data-frame-using-prepare_data" class="anchor"></a>Building data frame using <code>prepare_data</code>
</h2>
<p>We’ll already learned that <code>prepare_data</code> can join counts and covariates into a single data.frame. It can also compute offset through <code>compute_offset</code> and does so by default with <code>offset = "TSS"</code>, hence the <code>Offset</code> column in <code>trichoptera2</code>. You can change the offset method and provide additional arguments that will passed on to <code>compute_offset</code>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(<span class="kw"><a href="../reference/prepare_data.html">prepare_data</a></span>(trichoptera<span class="op">$</span>Abundance, </a>
<a class="sourceLine" id="cb18-2" data-line-number="2">             trichoptera<span class="op">$</span>Covariate, </a>
<a class="sourceLine" id="cb18-3" data-line-number="3">             <span class="dt">offset =</span> <span class="st">"RLE"</span>, <span class="dt">pseudocounts =</span> <span class="dv">1</span>))</a></code></pre></div>
<pre><code>## 'data.frame':    49 obs. of  9 variables:
##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr  "1" "2" "3" "4" ...
##   .. ..$ : chr  "Che" "Hyc" "Hym" "Hys" ...
##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...
##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...
##  $ Pressure     : num  998 1000 997 991 990 ...
##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...
##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...
##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...
##  $ Group        : Factor w/ 12 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Offset       : num  0.919 0.835 0.857 0.919 0.919 ...</code></pre>
<p>Different communities use different standard for the count data where samples are either or columns of the counts matrix. <code>prepare_data</code> uses heuristics to guess the direction of the counts matrix (or fail informatively doing so) and automatically transpose it if needed.</p>
<p>Finally, <code>prepare_data</code> enforces sample-consistency between the counts and the covariates and automatically trims away: - samples for which only covariates or only counts are available; - samples with no positive counts</p>
<p>For example, if we remove the first sample from the counts and the last one from the covariates, we end up with 49 - 2 = 47 samples left, as expected.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(<span class="kw"><a href="../reference/prepare_data.html">prepare_data</a></span>(trichoptera<span class="op">$</span>Abundance[<span class="op">-</span><span class="dv">1</span>, ], ## remove first sample</a>
<a class="sourceLine" id="cb20-2" data-line-number="2">                  trichoptera<span class="op">$</span>Covariate[<span class="op">-</span><span class="dv">49</span>,]  ## remove last sample</a>
<a class="sourceLine" id="cb20-3" data-line-number="3">                  ))</a></code></pre></div>
<pre><code>## [1] 47</code></pre>
</div>
<div id="importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom" class="anchor"></a>Importing data from biom and phyloseq objects using <code>prepare_data_from_[phyloseq|biom]</code>
</h2>
<p>Community composition data are quite popular in microbial ecology and usually stored in flat files using the <a href="http://biom-format.org/">biom format</a> and/or imported in R as phyloseq-class objects <span class="citation">(McMurdie <a href="#ref-phyloseq">2013</a>)</span> using the Bioconductor <a href="https://joey711.github.io/phyloseq/">phyloseq</a> package.</p>
<p>We provide helper functions to directly import data from a biom file (or biom-class object) and a phyloseq-class object.</p>
<div id="reading-from-a-biom-file" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-from-a-biom-file" class="anchor"></a>Reading from a biom file</h3>
<p>Note that the covariates <strong>must</strong> be stored in the biom object as they are automatically extracted, reading a biom without covariates results in an error.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(biomformat)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">## Biom file without covariates</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">biomfile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"min_dense_otu_table.biom"</span>, <span class="dt">package =</span> <span class="st">"biomformat"</span>)</a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="kw"><a href="../reference/prepare_data_from_biom.html">prepare_data_from_biom</a></span>(biomfile)</a></code></pre></div>
<pre><code>## Error in prepare_data_from_biom(biomfile): No covariates detected in biom. Consider:
## - extracting count data from biom with biom_data()
## - preparing a covariates data.frame
## - using prepare_data instead of prepare_data_from_biom</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">## Biom file with covariates, TSS offsets</a>
<a class="sourceLine" id="cb24-2" data-line-number="2">biomfile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"rich_dense_otu_table.biom"</span>, <span class="dt">package =</span> <span class="st">"biomformat"</span>)</a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(<span class="kw"><a href="../reference/prepare_data_from_biom.html">prepare_data_from_biom</a></span>(biomfile, <span class="dt">offset =</span> <span class="st">"TSS"</span>))</a></code></pre></div>
<pre><code>## 'data.frame':    6 obs. of  6 variables:
##  $ Abundance           : num [1:6, 1:5] 0 0 1 0 0 0 5 1 0 2 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr  "Sample1" "Sample2" "Sample3" "Sample4" ...
##   .. ..$ : chr  "GG_OTU_1" "GG_OTU_2" "GG_OTU_3" "GG_OTU_4" ...
##  $ BarcodeSequence     : chr  "CGCTTATCGAGA" "CATACCAGTAGC" "CTCTCTACCTGT" "CTCTCGGCCTGT" ...
##  $ LinkerPrimerSequence: chr  "CATGCTGCCTCCCGTAGGAGT" "CATGCTGCCTCCCGTAGGAGT" "CATGCTGCCTCCCGTAGGAGT" "CATGCTGCCTCCCGTAGGAGT" ...
##  $ BODY_SITE           : chr  "gut" "gut" "gut" "skin" ...
##  $ Description         : chr  "human gut" "human gut" "human gut" "human skin" ...
##  $ Offset              : num  7 3 4 6 5 2</code></pre>
</div>
<div id="reading-from-a-phyloseq-class-object" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-from-a-phyloseq-class-object" class="anchor"></a>Reading from a phyloseq-class object</h3>
<p>Note that the covariates <strong>must</strong> be stored in the phyloseq-class object as they are automatically extracted, importing a phyloseq object with no <code>sample_data</code> component results in an error.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(phyloseq)</a>
<a class="sourceLine" id="cb26-2" data-line-number="2">## phyloseq object without covariates</a>
<a class="sourceLine" id="cb26-3" data-line-number="3">physeq &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/phyloseq">phyloseq</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/otu_table-methods">otu_table</a></span>(trichoptera<span class="op">$</span>Abundance, <span class="dt">taxa_are_rows =</span> <span class="ot">FALSE</span>))</a>
<a class="sourceLine" id="cb26-4" data-line-number="4"><span class="kw"><a href="../reference/prepare_data_from_phyloseq.html">prepare_data_from_phyloseq</a></span>(physeq)</a></code></pre></div>
<pre><code>## Error in prepare_data_from_phyloseq(physeq): physeq should be a phyloseq object.</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">## phyloseq object with covariates, TSS offsets</a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"GlobalPatterns"</span>)</a>
<a class="sourceLine" id="cb28-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/str">str</a></span>(<span class="kw"><a href="../reference/prepare_data_from_phyloseq.html">prepare_data_from_phyloseq</a></span>(GlobalPatterns, <span class="dt">offset =</span> <span class="st">"TSS"</span>))</a></code></pre></div>
<pre><code>## 'data.frame':    26 obs. of  9 variables:
##  $ Abundance               : num [1:26, 1:19216] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr  "CL3" "CC1" "SV1" "M31Fcsw" ...
##   .. ..$ : chr  "549322" "522457" "951" "244423" ...
##  $ X.SampleID              : Factor w/ 26 levels "AQC1cm","AQC4cm",..: 5 4 21 14 11 15 12 9 16 13 ...
##  $ Primer                  : Factor w/ 26 levels "ILBC_01","ILBC_02",..: 1 2 3 4 5 6 7 8 9 10 ...
##  $ Final_Barcode           : Factor w/ 26 levels "AACGCA","AACTCG",..: 1 2 3 4 5 6 7 8 9 10 ...
##  $ Barcode_truncated_plus_T: Factor w/ 26 levels "AACTGT","ACAGGT",..: 24 13 3 20 9 5 17 7 19 11 ...
##  $ Barcode_full_length     : Factor w/ 26 levels "AGAGAGACAGG",..: 10 6 18 21 8 9 16 19 26 20 ...
##  $ SampleType              : Factor w/ 9 levels "Feces","Freshwater",..: 8 8 8 1 1 7 7 7 9 9 ...
##  $ Description             : Factor w/ 25 levels "Allequash Creek, 0-1cm depth",..: 4 5 20 14 11 15 12 9 16 13 ...
##  $ Offset                  : num  864077 1135457 697509 1543451 2076476 ...</code></pre>
</div>
</div>
<div id="mathematical-details-about-the-offsets" class="section level2">
<h2 class="hasAnchor">
<a href="#mathematical-details-about-the-offsets" class="anchor"></a>Mathematical details about the offsets</h2>
<p>We detail here the mathematical background behind the various offsets and the way they are computed. Note <span class="math inline">\(\mathbf{C} = (c_{ij})\)</span> the counts matrix where <span class="math inline">\(c_{ij}\)</span> is the count of species <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span>. Assume that there are <span class="math inline">\(J\)</span> species in total. The offset of sample <span class="math inline">\(i\)</span> is noted <span class="math inline">\(O_i\)</span> and computed in the following way.</p>
<div id="total-sum-scaling" class="section level3">
<h3 class="hasAnchor">
<a href="#total-sum-scaling" class="anchor"></a>Total Sum Scaling</h3>
<p>Offsets are simply the total counts of a sample: <span class="math display">\[
O_i = \sum_{j=1}^J c_{ij}
\]</span></p>
</div>
<div id="cumulative-sum-scaling" class="section level3">
<h3 class="hasAnchor">
<a href="#cumulative-sum-scaling" class="anchor"></a>Cumulative Sum Scaling</h3>
<p>Positive counts are used to compute sample-specific quantiles <span class="math inline">\(q_i^l\)</span> and cumulative sums <span class="math inline">\(s_i^l\)</span> defined as <span class="math display">\[
q_i^l = \min \{q \text{ such that } \sum_j 1_{c_{ij} \leq q} \geq l \sum_j 1_{c_{ij} &gt; 0} \} \qquad s_i^l = \sum_{j: c_{ij} \leq q_i^l} c_{ij}
\]</span> The sample-specific quantiles are then used to compute reference quantiles defined as <span class="math inline">\(q^l = \text{median} \{q^i_l\}\)</span> and median average deviation around the quantile <span class="math inline">\(q^l\)</span> as <span class="math inline">\(d^l = \text{median} |q_i^l - q^l|\)</span>. The method then searches for the smallest quantile <span class="math inline">\(l\)</span> for which it detects instability, defined as large relative increase in the <span class="math inline">\(d^l\)</span>. Formally, <span class="math inline">\(\hat{l}\)</span> is the smallest <span class="math inline">\(l\)</span> satisfying <span class="math inline">\(\frac{d^{l+1} - d^l}{d^l} \geq 0.1\)</span>. The scaling sample-specific offset are then chosen as: <span class="math display">\[
O_i = s_i^{\hat{l}} / \text{median}_i \{ s_i^{\hat{l}} \}
\]</span> Dividing by the median of the <span class="math inline">\(s_i^{\hat{l}}\)</span> ensures that offsets are centered around <span class="math inline">\(1\)</span> and compare sizes differences with respect to the reference sample. Note also that the reference quantiles <span class="math inline">\(q^l\)</span> can be computed using either the median (default, as in the original <span class="citation">Paulson et al. (<a href="#ref-CSS">2013</a>)</span> paper) or the mean, by specifying <code>reference = mean</code>, as implemented in <code>metagenomeseq</code>.</p>
</div>
<div id="relative-log-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#relative-log-expression" class="anchor"></a>Relative Log Expression</h3>
<p>A reference sample <span class="math inline">\((q_j)_j\)</span> is first built by computing the geometric means of each species count: <span class="math display">\[
q_j = \exp \left( \frac{1}{n} \sum_{i} \log(c_{ij})\right)
\]</span> Each sample is then compared to the reference sample to compute one ratio per species and the final offset <span class="math inline">\(O_i\)</span> is the median of those ratios: <span class="math display">\[
O_i = \text{median}_j \frac{c_{ij}}{q_j}
\]</span> The method fails when no species is shared across all sample (as all <span class="math inline">\(q_j\)</span> are then <span class="math inline">\(0\)</span>) or when a sample shares less than 50% of species with the reference (in which case the median of the ratios may be null or infinite). The problem can be alleviated by adding pseudocounts to the <span class="math inline">\(c_{ij}\)</span> with <code>pseudocounts = 1</code>.</p>
</div>
<div id="geometric-mean-of-pairwise-ratio" class="section level3">
<h3 class="hasAnchor">
<a href="#geometric-mean-of-pairwise-ratio" class="anchor"></a>Geometric Mean of Pairwise Ratio</h3>
<p>This method is similar to RLE but does create a reference sample. Instead, each sample is compared to each other to compute a median ratio (similar to RLE) <span class="math display">\[
r_{ii'} = {\text{median}}_{j: c_{ij}.c_{i'j} &gt; 0} \frac{c_{ij}}{c_{i'j}}
\]</span> The offset is then taken as the median of all the <span class="math inline">\(r_{ii'}\)</span>: <span class="math display">\[
O_i = \text{median}_{i' != i} r_{ii'}
\]</span> The method fails when there is only one sample in the data set or when a sample shares no species with any other.</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-DESeq2">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11 (10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106" class="uri">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-GMPR">
<p>Chen, Li, James Reeve, Lujun Zhang, Shengbing Huang, Xuefeng Wang, and Jun Chen. 2018. “GMPR: A Robust Normalization Method for Zero-Inflated Count Data with Application to Microbiome Sequencing Data.” <em>PeerJ</em> 6 (April): e4600. <a href="https://doi.org/10.7717/peerj.4600" class="uri">https://doi.org/10.7717/peerj.4600</a>.</p>
</div>
<div id="ref-phyloseq">
<p>McMurdie, Paul J. AND Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLoS ONE</em> 8 (4). Public Library of Science: e61217. <a href="https://doi.org/10.1371/journal.pone.0061217" class="uri">https://doi.org/10.1371/journal.pone.0061217</a>.</p>
</div>
<div id="ref-CSS">
<p>Paulson, Joseph N, O. Colin Stine, Héctor Corrada Bravo, and Mihai Pop. 2013. “Differential Abundance Analysis for Microbial Marker-Gene Surveys.” <em>Nat Methods</em> 10 (September): 1200–1202. <a href="https://doi.org/10.1038/nmeth.2658" class="uri">https://doi.org/10.1038/nmeth.2658</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>although a <code>data.frame</code> is technically a <code>list</code><a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#preliminaries">Preliminaries</a></li>
      <li><a href="#format-description">Format description</a></li>
      <li><a href="#computing-offsets">Computing offsets</a></li>
      <li><a href="#building-data-frame-using-prepare_data">Building data frame using <code>prepare_data</code></a></li>
      <li><a href="#importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom">Importing data from biom and phyloseq objects using <code>prepare_data_from_[phyloseq|biom]</code></a></li>
      <li><a href="#mathematical-details-about-the-offsets">Mathematical details about the offsets</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Julien Chiquet, Mahendra Mariadassou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
