<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data importation in PLNmodels • PLNmodels</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data importation in PLNmodels">
<meta property="og:description" content="PLNmodels">
<meta property="og:image" content="https://pln-team.github.io/PLNmodels/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PLNmodels</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.5-9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://pln-team.github.io/slideshow/slides">Slideshow on PLN</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Trichoptera.html">The trichoptera data set</a>
    </li>
    <li>
      <a href="../articles/Import_data.html">Data importation in PLNmodels</a>
    </li>
    <li>
      <a href="../articles/PLN.html">PLN: Multivariate Poisson regression</a>
    </li>
    <li>
      <a href="../articles/PLNPCA.html">PLNPCA: Dimension Reduction</a>
    </li>
    <li>
      <a href="../articles/PLNnetwork.html">PLNnetwork: sparse structure inference</a>
    </li>
    <li>
      <a href="../articles/PLNLDA.html">PLNLDA: Discriminant Analysis</a>
    </li>
    <li>
      <a href="../articles/PLNmixture.html">PLNmixture: Model-based clustering for counts</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pln-team/PLNmodels/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Import_data_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data importation in PLNmodels</h1>
                        <h4 class="author">PLN team</h4>
            
            <h4 class="date">2021-09-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pln-team/PLNmodels/blob/master/vignettes/Import_data.Rmd"><code>vignettes/Import_data.Rmd</code></a></small>
      <div class="hidden name"><code>Import_data.Rmd</code></div>

    </div>

    
    
<div id="preliminaries" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminaries" class="anchor"></a>Preliminaries</h2>
<p>This vignette documents the data format used in <strong>PLNmodel</strong> by <code>PLN</code> and its variants. It also shows how to create an object in the proper format for further analyses from (i) tabular data, (ii) biom-class objects and (iii) phyloseq-class objects.</p>
</div>
<div id="format-description" class="section level2">
<h2 class="hasAnchor">
<a href="#format-description" class="anchor"></a>Format description</h2>
<p>We illustrate the format using trichoptera data set, a full description of which can be found in <a href="Trichoptera.html">the corresponding vignette</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://pln-team.github.io/PLNmodels/">PLNmodels</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">)</span></code></pre></div>
<p>The trichoptera data set is a list made of two data frames: <code>Abundance</code> (hereafter referred to as the <em>counts</em>) and <code>Covariate</code> (hereafter the <em>covariates</em>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">trichoptera</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## List of 2
##  $ Abundance:'data.frame':   49 obs. of  17 variables:
##  $ Covariate:'data.frame':   49 obs. of  7 variables:</code></pre>
<p>The covariates include, among others, the wind, pressure and humidity.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Temperature"   "Wind"          "Pressure"      "Humidity"     
## [5] "Cloudiness"    "Precipitation" "Group"</code></pre>
<p>In the PLN framework, we model the counts from the covariates, let’s say wind and pressure, using a Poisson Log-Normal model. Most models in R use the so-called <em>formula interface</em> and it would thus be natural to write something like</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/PLN.html">PLN</a></span><span class="op">(</span><span class="va">Abundance</span> <span class="op">~</span> <span class="va">Wind</span> <span class="op">+</span> <span class="va">Pressure</span>, data <span class="op">=</span> <span class="va">trichoptera</span><span class="op">)</span></code></pre></div>
<p>Unfortunately and unlike many generalized linear models, the response in PLN is intrinsically <strong>multivariate</strong>: it has 17 dimensions in our example. The left hand side (LHS) must encode a multivariate response across multiple samples, using a 2D-array (e.g. a matrix or a data frame).</p>
<p>We must therefore prepare a data structure where <code>Abundance</code> refers to a count <em>matrix</em> whereas <code>Wind</code> and <code>Pressure</code> refer to <em>vectors</em> before feeding it to <code>PLN</code>. That’s the purpose of <code>prepare_data</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trichoptera2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span>counts     <span class="op">=</span> <span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, 
                             covariates <span class="op">=</span> <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">trichoptera2</span><span class="op">)</span></code></pre></div>
<pre><code>## 'data.frame':    49 obs. of  9 variables:
##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr [1:49] "1" "2" "3" "4" ...
##   .. ..$ : chr [1:17] "Che" "Hyc" "Hym" "Hys" ...
##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...
##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...
##  $ Pressure     : num  998 1000 997 991 990 ...
##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...
##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...
##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...
##  $ Group        : Factor w/ 12 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Offset       : num  29 13 38 192 79 18 8 34 12 4 ...</code></pre>
<p>If you look carefully, you can notice a few difference between <code>trichoptera</code> and <code>trichoptera2</code>:</p>
<ul>
<li>the first is a <code>list</code> whereas the second is a <code>data.frame</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>;</li>
<li>
<code>Abundance</code> is a matrix-column of <code>trichoptera2</code> that you can extract using the usual functions <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> and <code><a href="https://rdrr.io/r/base/Extract.html">[[</a></code> to retrieve the count matrix;</li>
<li>
<code>trichoptera2</code> has an additional <code>Offset</code> column (more on that later).</li>
</ul>
</div>
<div id="computing-offsets" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-offsets" class="anchor"></a>Computing offsets</h2>
<p>It is common practice when modeling count data to introduce an offset term to control for different sampling efforts, exposures, baselines, etc. The <em>proper way</em> to compute sample-specific offsets in still debated and may vary depending on the field. There are nevertheless a few popular methods:</p>
<ul>
<li>Total Sum Scaling (TSS), where the offset of a sample is the total count in that sample</li>
<li>Cumulative Sum Scaling (CSS), introduced in <span class="citation">(Paulson et al. <a href="#ref-CSS" role="doc-biblioref">2013</a>)</span>, where the offset of a sample if the cumulative sum of counts in that sample, up to a quantile determined in a data driven way.</li>
<li>Relative Log-Expression (RLE), implemented in <span class="citation">(Anders and Huber <a href="#ref-DESeq2" role="doc-biblioref">2010</a>)</span>, where all samples are used to compute a reference sample, each sample is compared to the reference sample using log-ratios and the offset is the median log-ratio.</li>
<li>Geometric Mean of Pairwise Ratio (GMPR), introduced in <span class="citation">(Chen et al. <a href="#ref-GMPR" role="doc-biblioref">2018</a>)</span> where each sample is compared to each other to compute a median log-ratio and the offset of a sample is the geometric means of those pairwise ratios.</li>
<li>Wrench, introduced in <span class="citation">(Kumar et al. <a href="#ref-Kumar2018" role="doc-biblioref">2018</a>)</span> and fully implemented in the <a href="http://bioconductor.org/packages/release/bioc/html/Wrench.html">Wrench package</a>, where all samples are used to compute reference proportions and each sample is compared to the reference using ratios (and <strong>not log-ratios</strong>) of proportions to compute compositional correction factors. In that case, the offset is the product of (geometrically centered) compositional factors and (geometrically centered) depths.</li>
</ul>
<p>Each of these offset be computed from a counts matrix using the <code>compute_offset</code> function and changing its <code>offset</code> argument:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## same as compute_offset(trichoptera$Abundance, offset = "TSS")</span>
<span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span><span class="op">)</span> </code></pre></div>
<pre><code>##    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
##   29   13   38  192   79   18    8   34   12    4    4    3   49   33  600  172 
##   17   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32 
##   58   51   56  127   35   13   17    3   27   40   44    8    9 1599 2980   88 
##   33   34   35   36   37   38   39   40   41   42   43   44   45   46   47   48 
##  135  327   66   90   63   15   14   20   70   53   95   43   62  149   16   31 
##   49 
##   86</code></pre>
<p>In this particular example, the counts are too sparse and sophisticated offset methods all fail (numeric output hidden)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"CSS"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in offset_function(counts, ...): Some samples only have 1 positive
## values. Can't compute quantiles and fall back to TSS normalization</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"RLE"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in offset_function(counts, ...): Because of high sparsity, some samples
## have null or infinite offset.</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"GMPR"</span><span class="op">)</span></code></pre></div>
<p>We can mitigate this problem for the RLE offset by adding pseudocounts to the counts although doing so has its own drawbacks.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"RLE"</span>, pseudocounts <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>##         1         2         3         4         5         6         7         8 
## 0.9186270 0.8349121 0.8570257 0.9186270 0.9186270 0.8349121 0.8192245 0.8570257 
##         9        10        11        12        13        14        15        16 
## 0.7322797 0.7322797 0.6321923 0.6321923 0.9240361 0.9186270 1.3788037 0.9240361 
##        17        18        19        20        21        22        23        24 
## 0.9186270 0.9240361 0.9240361 1.7140514 0.8908577 0.8570257 0.8349121 0.6321923 
##        25        26        27        28        29        30        31        32 
## 0.9240361 0.9240361 0.9186270 0.8570257 0.8349121 2.7721084 3.2934218 0.9584503 
##        33        34        35        36        37        38        39        40 
## 1.0406547 0.9584503 0.9584503 0.9186270 0.9584503 0.8908577 0.8349121 0.7322797 
##        41        42        43        44        45        46        47        48 
## 0.9240361 0.9240361 0.9584503 0.9584503 1.2643846 1.7140514 0.8233555 0.8908577 
##        49 
## 0.9186270</code></pre>
<p>A better solution consists in using only positive counts to compute the offsets:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"RLE"</span>, type <span class="op">=</span> <span class="st">"poscounts"</span><span class="op">)</span></code></pre></div>
<pre><code>##         1         2         3         4         5         6         7         8 
## 0.5631099 0.9462046 0.8299806 0.9462046 0.6460415 0.5756774 0.6308031 0.4947789 
##         9        10        11        12        13        14        15        16 
## 0.7988730 0.3574190 0.2207270 0.1260525 0.8051707 0.7289732 1.5591914 1.1850090 
##        17        18        19        20        21        22        23        24 
## 0.6389431 1.0000000 1.4606879 1.8224330 0.8403498 0.3942114 0.4058586 0.1997183 
##        25        26        27        28        29        30        31        32 
## 0.6286560 0.5631099 0.8777257 0.3440303 0.2504662 2.7086423 2.5800932 1.4584997 
##        33        34        35        36        37        38        39        40 
## 2.6050843 4.5898981 1.2185072 1.2616062 0.8823673 0.6250713 0.3950030 0.5631099 
##        41        42        43        44        45        46        47        48 
## 1.4511384 1.0000000 0.9462046 1.0882448 1.0000000 1.0631099 0.4621924 0.7900060 
##        49 
## 1.0672361</code></pre>
<p>Finally, we can use wrench to compute the offsets:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/compute_offset.html">compute_offset</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="st">"Wrench"</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1]  0.41269451  0.17385897  0.31925785  0.70682391  0.44749382  0.20676142
##  [7]  0.10226641  0.21204241  0.09228293  0.03919178  0.03946910  0.02295077
## [13] 24.18096281  1.54038084  3.95187144  0.80172207 10.52136472  4.55854738
## [19]  0.46968859 51.88528943  0.32803605  0.14764973  4.13635353  0.03152826
## [25]  1.63626578  0.53769290  0.38482355  0.62689594  0.08030685 83.87692476
## [31] 33.40810265  0.75508731  1.05809625  1.18385013  0.58657751  0.43005272
## [37]  4.79925224  0.18381986  0.16078594  0.18490773  4.25999137  5.46689591
## [43]  0.60139850  7.56594658  9.89766712 33.73727074  0.74264459 29.13747796
## [49] 47.12017975</code></pre>
</div>
<div id="building-data-frame-using-prepare_data" class="section level2">
<h2 class="hasAnchor">
<a href="#building-data-frame-using-prepare_data" class="anchor"></a>Building data frame using <code>prepare_data</code>
</h2>
<p>We’ll already learned that <code>prepare_data</code> can join counts and covariates into a single data.frame. It can also compute offset through <code>compute_offset</code> and does so by default with <code>offset = "TSS"</code>, hence the <code>Offset</code> column in <code>trichoptera2</code>. You can change the offset method and provide additional arguments that will passed on to <code>compute_offset</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, 
             <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span>, 
             offset <span class="op">=</span> <span class="st">"RLE"</span>, pseudocounts <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 'data.frame':    49 obs. of  9 variables:
##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : chr [1:49] "1" "2" "3" "4" ...
##   .. ..$ : chr [1:17] "Che" "Hyc" "Hym" "Hys" ...
##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...
##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...
##  $ Pressure     : num  998 1000 997 991 990 ...
##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...
##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...
##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...
##  $ Group        : Factor w/ 12 levels "1","2","3","4",..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Offset       : num  0.919 0.835 0.857 0.919 0.919 ...</code></pre>
<p>Different communities use different standard for the count data where samples are either or columns of the counts matrix. <code>prepare_data</code> uses heuristics to guess the direction of the counts matrix (or fail informatively doing so) and automatically transpose it if needed.</p>
<p>Finally, <code>prepare_data</code> enforces sample-consistency between the counts and the covariates and automatically trims away: - samples for which only covariates or only counts are available; - samples with no positive counts</p>
<p>For example, if we remove the first sample from the counts and the last one from the covariates, we end up with 49 - 2 = 47 samples left, as expected.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>, <span class="op">]</span>, <span class="co">## remove first sample</span>
                  <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">[</span><span class="op">-</span><span class="fl">49</span>,<span class="op">]</span>  <span class="co">## remove last sample</span>
                  <span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 47</code></pre>
</div>
<div id="importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom" class="anchor"></a>Importing data from biom and phyloseq objects using <code>prepare_data_from_[phyloseq|biom]</code>
</h2>
<p>Community composition data are quite popular in microbial ecology and usually stored in flat files using the <a href="http://biom-format.org/">biom format</a> and/or imported in R as phyloseq-class objects <span class="citation">(McMurdie <a href="#ref-phyloseq" role="doc-biblioref">2013</a>)</span> using the Bioconductor <a href="https://joey711.github.io/phyloseq/">phyloseq</a> package.</p>
<!-- We provide helper functions to directly import data from a biom file (or biom-class object) and a phyloseq-class object.  -->
<p>We show here how to import data from a biom file (or biom-class object) and form a phyloseq-class object.</p>
<div id="reading-from-a-biom-file" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-from-a-biom-file" class="anchor"></a>Reading from a biom file</h3>
<p>Reading from a biom file requires the bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/biomformat.html">biomformat</a>. This package is <strong>not</strong> a standard dependency of PLNmodels and needs to be installed separately.</p>
<p>You can easily prepare your data from a biom file using the following steps:</p>
<ul>
<li>read your biom file with <code><a href="https://rdrr.io/pkg/biomformat/man/read_biom.html">biomformat::read_biom()</a></code>
</li>
<li>extract the count table with <code><a href="https://rdrr.io/pkg/biomformat/man/biom_data-methods.html">biomformat::biom_data()</a></code>
</li>
<li>extract the covariates with <code><a href="https://rdrr.io/pkg/biomformat/man/sample_metadata-methods.html">biomformat::sample_metadata()</a></code> (or build your own)</li>
<li>feed them to <code>prepare_data</code>
</li>
</ul>
<p>as illustrated below: <!-- Note that the covariates **must** be stored in the biom object as they are automatically extracted, reading a biom without covariates results in an error.  --></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## If biomformat is not installed, uncomment the following lines</span>
<span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE)) {</span>
<span class="co">#   install.packages("BiocManager")</span>
<span class="co"># }</span>
<span class="co"># BiocManager::install("biomformat")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joey711/biomformat/">biomformat</a></span><span class="op">)</span>
<span class="va">biomfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rich_dense_otu_table.biom"</span>, package <span class="op">=</span> <span class="st">"biomformat"</span><span class="op">)</span>
<span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">biomformat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomformat/man/read_biom.html">read_biom</a></span><span class="op">(</span><span class="va">biomfile</span><span class="op">)</span>
<span class="co">## extract counts</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu">biomformat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomformat/man/biom_data-methods.html">biom_data</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span>
<span class="co">## extract covariates (or prepare your own)</span>
<span class="va">covariates</span> <span class="op">&lt;-</span> <span class="fu">biomformat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biomformat/man/sample_metadata-methods.html">sample_metadata</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span>
<span class="co">## prepare data</span>
<span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, covariates <span class="op">=</span> <span class="va">covariates</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">my_data</span><span class="op">)</span></code></pre></div>
</div>
<div id="reading-from-a-phyloseq-class-object" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-from-a-phyloseq-class-object" class="anchor"></a>Reading from a phyloseq-class object</h3>
<p>Likewise, preparing data from a phyloseq-class object requires the bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html">phyloseq</a>. This package is <strong>not</strong> a standard dependency of PLNmodels and needs to be installed separately.</p>
<p>You can easily prepare your data from a phyloseq object using the following steps:</p>
<ul>
<li>extract the count table with <code><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">phyloseq::otu_table()</a></code>
</li>
<li>extract the covariates with <code><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">phyloseq::sample_data()</a></code> (or build your own)</li>
<li>feed them to <code>prepare_data</code>
</li>
</ul>
<p>as illustrated below:</p>
<!-- Note that the covariates **must** be stored in the phyloseq-class object as they are automatically extracted, importing a phyloseq object with no `sample_data` component results in an error.  -->
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## If biomformat is not installed, uncomment the following lines</span>
<span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE)) {</span>
<span class="co">#   install.packages("BiocManager")</span>
<span class="co"># }</span>
<span class="co"># BiocManager::install("phyloseq")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"enterotype"</span><span class="op">)</span>
<span class="co">## extract counts</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span>
<span class="co">## extract covariates (or prepare your own)</span>
<span class="va">covariates</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">enterotype</span><span class="op">)</span>
<span class="co">## prepare data</span>
<span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, covariates <span class="op">=</span> <span class="va">covariates</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">my_data</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="mathematical-details-about-the-offsets" class="section level2">
<h2 class="hasAnchor">
<a href="#mathematical-details-about-the-offsets" class="anchor"></a>Mathematical details about the offsets</h2>
<p>We detail here the mathematical background behind the various offsets and the way they are computed. Note <span class="math inline">\(\mathbf{Y} = (Y_{ij})\)</span> the counts matrix where <span class="math inline">\(Y_{ij}\)</span> is the count of species <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span>. Assume that there are <span class="math inline">\(p\)</span> species and <span class="math inline">\(n\)</span> samples in total. The offset of sample <span class="math inline">\(i\)</span> is noted <span class="math inline">\(O_i\)</span> and computed in the following way.</p>
<div id="total-sum-scaling" class="section level3">
<h3 class="hasAnchor">
<a href="#total-sum-scaling" class="anchor"></a>Total Sum Scaling</h3>
<p>Offsets are simply the total counts of a sample (frequently called depths in the metabarcoding literature): <span class="math display">\[
O_i = \sum_{j=1}^p Y_{ij}
\]</span></p>
</div>
<div id="cumulative-sum-scaling" class="section level3">
<h3 class="hasAnchor">
<a href="#cumulative-sum-scaling" class="anchor"></a>Cumulative Sum Scaling</h3>
<p>Positive counts are used to compute sample-specific quantiles <span class="math inline">\(q_i^l\)</span> and cumulative sums <span class="math inline">\(s_i^l\)</span> defined as <span class="math display">\[
q_i^l = \min \{q \text{ such that } \sum_j 1_{Y_{ij} \leq q} \geq l \sum_j 1_{Y_{ij} &gt; 0} \} \qquad s_i^l = \sum_{j: Y_{ij} \leq q_i^l} Y_{ij}
\]</span> The sample-specific quantiles are then used to compute reference quantiles defined as <span class="math inline">\(q^l = \text{median} \{q^i_l\}\)</span> and median average deviation around the quantile <span class="math inline">\(q^l\)</span> as <span class="math inline">\(d^l = \text{median} |q_i^l - q^l|\)</span>. The method then searches for the smallest quantile <span class="math inline">\(l\)</span> for which it detects instability, defined as large relative increase in the <span class="math inline">\(d^l\)</span>. Formally, <span class="math inline">\(\hat{l}\)</span> is the smallest <span class="math inline">\(l\)</span> satisfying <span class="math inline">\(\frac{d^{l+1} - d^l}{d^l} \geq 0.1\)</span>. The scaling sample-specific offset are then chosen as: <span class="math display">\[
O_i = s_i^{\hat{l}} / \text{median}_i \{ s_i^{\hat{l}} \}
\]</span> Dividing by the median of the <span class="math inline">\(s_i^{\hat{l}}\)</span> ensures that offsets are centered around <span class="math inline">\(1\)</span> and compare sizes differences with respect to the reference sample. Note also that the reference quantiles <span class="math inline">\(q^l\)</span> can be computed using either the median (default, as in the original <span class="citation">Paulson et al. (<a href="#ref-CSS" role="doc-biblioref">2013</a>)</span> paper) or the mean, by specifying <code>reference = mean</code>, as implemented in <code>metagenomeseq</code>.</p>
</div>
<div id="relative-log-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#relative-log-expression" class="anchor"></a>Relative Log Expression</h3>
<p>A reference sample <span class="math inline">\((q_j)_j\)</span> is first built by computing the geometric means of each species count: <span class="math display">\[
q_j = \exp \left( \frac{1}{n} \sum_{i} \log(Y_{ij})\right)
\]</span> Each sample is then compared to the reference sample to compute one ratio per species and the final offset <span class="math inline">\(O_i\)</span> is the median of those ratios: <span class="math display">\[
O_i = \text{median}_j \frac{Y_{ij}}{q_j}
\]</span> The method fails when no species is shared across all sample (as all <span class="math inline">\(q_j\)</span> are then <span class="math inline">\(0\)</span>) or when a sample shares less than 50% of species with the reference (in which case the median of the ratios may be null or infinite). The problem can be alleviated by adding pseudocounts to the <span class="math inline">\(c_{ij}\)</span> with <code>pseudocounts = 1</code> or using positive counts in the computations (<code>type = "poscounts"</code>)</p>
</div>
<div id="geometric-mean-of-pairwise-ratio" class="section level3">
<h3 class="hasAnchor">
<a href="#geometric-mean-of-pairwise-ratio" class="anchor"></a>Geometric Mean of Pairwise Ratio</h3>
<p>This method is similar to RLE but does create a reference sample. Instead, each sample is compared to each other to compute a median ratio (similar to RLE) <span class="math display">\[
r_{ii'} = {\text{median}}_{j: Y_{ij}.Y_{i'j} &gt; 0} \frac{Y_{ij}}{Y_{i'j}}
\]</span> The offset is then taken as the median of all the <span class="math inline">\(r_{ii'}\)</span>: <span class="math display">\[
O_i = \text{median}_{i' != i} r_{ii'}
\]</span> The method fails when there is only one sample in the data set or when a sample shares no species with any other.</p>
</div>
<div id="wrench-normalisation" class="section level3">
<h3 class="hasAnchor">
<a href="#wrench-normalisation" class="anchor"></a>Wrench normalisation</h3>
<p>This method is fully detailed in <span class="citation">Kumar et al. (<a href="#ref-Kumar2018" role="doc-biblioref">2018</a>)</span> and we only provide a barebone implementation corresponding to the defaults parameters of <code><a href="https://rdrr.io/pkg/Wrench/man/wrench.html">Wrench::wrench()</a></code>. Assume that samples belong to <span class="math inline">\(K\)</span> discrete groups and note <span class="math inline">\(g_i\)</span> the group of sample <span class="math inline">\(i\)</span>. Wrench is based on the following (simplified) log-normal model for counts: <span class="math display">\[
Y_{ij} \sim \pi_{ij} \delta_0 + (1 - \pi_{ij})\log\mathcal{N}(\mu_{ij}, \sigma^2_j)
\]</span> where the <span class="math inline">\(Y_{ij}\)</span> are independent and the mean <span class="math inline">\(\mu_{ij}\)</span> is decomposed as: <span class="math display">\[
\mu_{ij} = \underbrace{\log{p_{0j}}}_{\text{log-ref. prop.}} +  \underbrace{\log{d_i}}_{\text{log-depth}} + \underbrace{\log{\zeta_{0g_i}}}_{\text{log effect of group } g_i} + \underbrace{a_{i}}_{\text{(f|m)ixed effect}} + \underbrace{b_{ij}}_{\text{mixed effects}}
\]</span> where the random effects are independents centered gaussian and the depths is the total sum of counts: <span class="math display">\[
\begin{align*}
d_i &amp; = \sum_{j=1}^p c_{ij} \\
b_{ij} &amp; \sim \mathcal{N}(0, \eta^2_{g_i}) \\
\end{align*}
\]</span></p>
<p>The <strong>net</strong> log fold change <span class="math inline">\(\theta_{ij}\)</span> of the <strong>proportion ratio</strong> <span class="math inline">\(r_{ij} = c_{ij} / d_i p_{0j}\)</span> of species <span class="math inline">\(j\)</span> relative to the reference is <span class="math inline">\(\log(\theta_{ij}) \overset{\Delta}{=} \mathbb{E}[\log(r_{ij}) | a_i, b_{ij}] = \log{\zeta_{0g_i}} + a_i + b_{ij}\)</span>. We can decompose it as <span class="math inline">\(\theta_{ij} = \Lambda_i^{-1} v_{ij}\)</span> where <span class="math inline">\(\Lambda_i^{-1}\)</span> is the <em>compositional correction factor</em> and <span class="math inline">\(v_{ij}\)</span> is the fold change of <strong>true abundances</strong>.</p>
<p>With the above notations, the net fold change compounds both the fold change of true abundances and the compositional correction factors. With the assumption that the <span class="math inline">\(b_{ij}\)</span> are centered, <span class="math inline">\(\log(\hat{\Lambda}_i)\)</span> can be estimated through a robust average of the <span class="math inline">\(\hat{\theta}_{ij}\)</span>, which can themselves be computed from the log-ratio of proportions.</p>
<p>We detail here how the different parameters and/or effects are estimated.</p>
<ul>
<li>The reference proportions <span class="math inline">\(p_{0j}\)</span> are constructed as averages of the sample proportions <span class="math inline">\(p_{ij}\)</span> and the ratio are derived from both quantities <span class="math display">\[
p_{ij} = \frac{Y_{ij}}{\sum_{j=1}^p Y_{ij}} \qquad p_{0j} = \frac{1}{n} \sum_{i=1}^n p_{ij} \qquad r_{ij} = \frac{p_{ij}}{p_{0j}}
\]</span>
</li>
<li>The probabilities of absence <span class="math inline">\(\pi_{ij}\)</span> are estimated by fitting the following Bernoulli models: <span class="math display">\[
1_{\{Y_{ij} = 0\}} \sim \mathcal{B}(\pi_{j}^{d_i})
\]</span> and setting <span class="math inline">\(\hat{\pi}_{ij} = \hat{\pi}^{d_i}\)</span>
</li>
<li>The species variances <span class="math inline">\(\sigma^2_j\)</span> are estimated by fitting the following linear model (with no zero-inflation component) <span class="math display">\[
\log Y_{ij} \sim \log(d_i) + \mu_{g_i} + \mathcal{N}(0, \sigma^2_j)
\]</span> Note that in the original <code><a href="https://rdrr.io/pkg/Wrench/man/wrench.html">Wrench::wrench()</a></code>, the log depth <span class="math inline">\(\log(d_i)\)</span> is used as predictor but I believe it makes more sense to use it an offset.</li>
<li>set the group proportions <span class="math inline">\(p_{gj}\)</span> and group ratios <span class="math inline">\(r_{gj}\)</span> to: <span class="math display">\[
p_{gj} = \frac{\sum_{i : g_i = g} Y_{ij}}{\sum_{j, i : g_i = g} Y_{ij}} \qquad r_{gj} = \frac{p_{gj}}{p_{0j}}
\]</span>
</li>
<li>Estimate the location and dispersion parameters as: <span class="math display">\[
\hat{\zeta}_{0g} = \frac{\sum_{j=1}^p r_{gj}}{p} 
\qquad
\log{r_{g.}} = \frac{\sum_{j: r_{gj} \neq 1} \log{r_{gj}}}{\sum_{j: r_{gj} \neq 0} 1}   
\qquad
\hat{\eta}_{g}^2 = \frac{\sum_{j: r_{gj} \neq 1} (\log{r_{gj}} - \log{r_{g.}})^2}{\sum_{j: r_{gj} \neq 0} 1}
\]</span>
</li>
<li>Estimate the mixed effects as shrunken (and scaled) averages of the ratios <span class="math display">\[
\hat{a}_i = \frac{\sum_{j = 1}^p \frac{1}{\hat{\eta}^2_{g_i} + \hat{\sigma}^2_j} (\log{r_{ij}} - \log{\hat{\zeta}_{0g_i}})}{\sum_{j = 1}^p \frac{1}{\hat{\eta}^2_{g_i} + \hat{\sigma}^2_j}}
\qquad 
\hat{b}_{ij} = \frac{\hat{\eta}^2_{g_i}}{\hat{\eta}^2_{g_i} + \hat{\sigma}^2_j} \left( \log{r_{ij}} - \log\hat{\zeta}_{0g_i} - \hat{a}_i\right)
\]</span>
</li>
<li>Estimate the regularized ratios as: <span class="math display">\[
\hat{\theta}_{ij} = \exp\left( \log\hat{\zeta}_{0g_i} + \hat{a}_i + \hat{b}_{ij} \right)
\]</span>
</li>
<li>Estimate the compositional correction factors as (weighted) means of the regularized (and possibly corrected) ratios: <span class="math display">\[
\hat{\Lambda}_i = 
\begin{cases}
\sum_{j = 1}^p \hat{\theta}_{ij} \bigg/ p&amp; \text{ if type = "simple"} \\
\sum_{j = 1}^p \hat{\theta}_{ij} e^{-\hat{\sigma}_j^2 / 2} /  w_{ij} \bigg/ \sum_{j=1}^p 1/w_{ij}  &amp; \text{ if type = "wrench"} \\
\end{cases}
\]</span> where <span class="math inline">\(w_{ij} = (1 - \hat{\pi}_{ij})(\hat{\pi}_{ij} + e^{\hat{\sigma}_j^2 + \hat{\eta}_i^2} - 1)\)</span>. The correction term <span class="math inline">\(e^{\hat{\sigma}_j^2 / 2}\)</span> arises from the relation <span class="math inline">\(\mathbb{E}[r_{ij} | r_{ij} &gt; 0] = \theta_{ij} e^{\sigma_j^2/2}\)</span> and the weight <span class="math inline">\(w_{ij}\)</span> are marginal variances: <span class="math inline">\(\mathbb{V}[r_{ij}] = w_{ij}\)</span>.</li>
</ul>
<p>The offsets are then the product of compositional correction factors and depths: <span class="math display">\[
O_i = \frac{\hat{\Lambda}_i}{(\prod_{i = 1}^n \hat{\Lambda}_i)^{1/n}} \times \frac{d_i}{(\prod_{i = 1}^n d_i)^{1/n}}
\]</span></p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-DESeq2">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11 (10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-GMPR">
<p>Chen, Li, James Reeve, Lujun Zhang, Shengbing Huang, Xuefeng Wang, and Jun Chen. 2018. “GMPR: A Robust Normalization Method for Zero-Inflated Count Data with Application to Microbiome Sequencing Data.” <em>PeerJ</em> 6 (April): e4600. <a href="https://doi.org/10.7717/peerj.4600">https://doi.org/10.7717/peerj.4600</a>.</p>
</div>
<div id="ref-Kumar2018">
<p>Kumar, M. Senthil, Eric V. Slud, Kwame Okrah, Stephanie C. Hicks, Sridhar Hannenhalli, and Héctor Corrada Bravo. 2018. “Analysis and Correction of Compositional Bias in Sparse Sequencing Count Data.” <em>BMC Genomics</em> 19 (1). <a href="https://doi.org/10.1186/s12864-018-5160-5">https://doi.org/10.1186/s12864-018-5160-5</a>.</p>
</div>
<div id="ref-phyloseq">
<p>McMurdie, Paul J. AND Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLoS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217">https://doi.org/10.1371/journal.pone.0061217</a>.</p>
</div>
<div id="ref-CSS">
<p>Paulson, Joseph N, O. Colin Stine, Héctor Corrada Bravo, and Mihai Pop. 2013. “Differential Abundance Analysis for Microbial Marker-Gene Surveys.” <em>Nat Methods</em> 10 (September): 1200–1202. <a href="https://doi.org/10.1038/nmeth.2658">https://doi.org/10.1038/nmeth.2658</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>although a <code>data.frame</code> is technically a <code>list</code><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Julien Chiquet, Mahendra Mariadassou, François Gindraud.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
