<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Supervized classification of multivariate count table with the Poisson discriminant Analysis • PLNmodels</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Supervized classification of multivariate count table with the Poisson discriminant Analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PLNmodels</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://pln-team.github.io/slideshow/slides" class="external-link">Slideshow on PLN</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Trichoptera.html">The trichoptera data set</a>
    </li>
    <li>
      <a href="../articles/Import_data.html">Data importation in PLNmodels</a>
    </li>
    <li>
      <a href="../articles/PLN.html">PLN: Multivariate Poisson regression</a>
    </li>
    <li>
      <a href="../articles/PLNPCA.html">PLNPCA: Dimension Reduction</a>
    </li>
    <li>
      <a href="../articles/PLNnetwork.html">PLNnetwork: sparse structure inference</a>
    </li>
    <li>
      <a href="../articles/PLNLDA.html">PLNLDA: Discriminant Analysis</a>
    </li>
    <li>
      <a href="../articles/PLNmixture.html">PLNmixture: Model-based clustering for counts</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pln-team/PLNmodels/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Supervized classification of multivariate count
table with the Poisson discriminant Analysis</h1>
                        <h4 data-toc-skip class="author">PLN team</h4>
            
            <h4 data-toc-skip class="date">2025-03-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pln-team/PLNmodels/blob/master/vignettes/PLNLDA.Rmd" class="external-link"><code>vignettes/PLNLDA.Rmd</code></a></small>
      <div class="hidden name"><code>PLNLDA.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>This vignette illustrates the standard use of the <code>PLNLDA</code>
function and the methods accompanying the R6 Classes <code>PLNLDA</code>
and <code>PLNLDAfit</code>.</p>
<div class="section level3">
<h3 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h3>
<p>The packages required for the analysis are <strong>PLNmodels</strong>
plus some others for data manipulation and representation:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pln-team.github.io/PLNmodels/">PLNmodels</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-set">Data set<a class="anchor" aria-label="anchor" href="#data-set"></a>
</h3>
<p>We illustrate our point with the trichoptera data set, a full
description of which can be found in <a href="Trichoptera.html">the
corresponding vignette</a>. Data preparation is also detailed in <a href="Import_data.html">the specific vignette</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">)</span></span>
<span><span class="va">trichoptera</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span><span class="va">trichoptera</span><span class="op">$</span><span class="va">Abundance</span>, <span class="va">trichoptera</span><span class="op">$</span><span class="va">Covariate</span><span class="op">)</span></span></code></pre></div>
<p>The <code>trichoptera</code> data frame stores a matrix of counts
(<code>trichoptera$Abundance</code>), a matrix of offsets
(<code>trichoptera$Offset</code>) and some vectors of covariates
(<code>trichoptera$Wind</code>, <code>trichoptera$Temperature</code>,
etc.) In the following, we’re particularly interested in the
<code>trichoptera$Group</code> <strong>discrete</strong> covariate which
corresponds to disjoint time spans during which the catching took place.
The correspondence between group label and time spans is:</p>
<table class="table">
<thead><tr class="header">
<th align="center">Label</th>
<th align="center">Number.of.Consecutive.Nights</th>
<th align="center">Date</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td align="center">12</td>
<td align="center">June 59</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td align="center">5</td>
<td align="center">June 59</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td align="center">5</td>
<td align="center">June 59</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td align="center">4</td>
<td align="center">June 59</td>
</tr>
<tr class="odd">
<td align="center">5</td>
<td align="center">4</td>
<td align="center">July 59</td>
</tr>
<tr class="even">
<td align="center">6</td>
<td align="center">1</td>
<td align="center">June 59</td>
</tr>
<tr class="odd">
<td align="center">7</td>
<td align="center">3</td>
<td align="center">June 60</td>
</tr>
<tr class="even">
<td align="center">8</td>
<td align="center">4</td>
<td align="center">June 60</td>
</tr>
<tr class="odd">
<td align="center">9</td>
<td align="center">5</td>
<td align="center">June 60</td>
</tr>
<tr class="even">
<td align="center">10</td>
<td align="center">4</td>
<td align="center">June 60</td>
</tr>
<tr class="odd">
<td align="center">11</td>
<td align="center">1</td>
<td align="center">June 60</td>
</tr>
<tr class="even">
<td align="center">12</td>
<td align="center">1</td>
<td align="center">July 60</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="mathematical-background">Mathematical background<a class="anchor" aria-label="anchor" href="#mathematical-background"></a>
</h3>
<p>In the vein of <span class="citation">Fisher (<a href="#ref-Fi36">1936</a>)</span> and <span class="citation">Rao (<a href="#ref-Rao48">1948</a>)</span>, we introduce a multi-class LDA model
for multivariate count data which is a variant of the Poisson Lognormal
model of <span class="citation">Aitchison and Ho (<a href="#ref-AiH89">1989</a>)</span> (see <a href="PLN.html">the PLN
vignette</a> as a reminder). Indeed, it can viewed as a PLN model with a
discrete group structure in the latent Gaussian space.</p>
<p>This PLN-LDA model can be written in a hierarchical framework where a
sample of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-dimensional
observation vectors
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐘</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{Y}_i</annotation></semantics></math>
is related to some
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-dimensional
vectors of latent variables
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐙</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{Z}_i</annotation></semantics></math>
and a discrete structure with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
groups in the following way:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">group structure </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>𝛍</mi><mi>i</mi></msub><mo>=</mo><msub><mi>μ</mi><msub><mi>g</mi><mi>i</mi></msub></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mi>g</mi><mi>i</mi></msub><mo>∈</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>K</mi><mo stretchy="false" form="postfix">}</mo></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">latent space </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>𝐙</mi><mi>i</mi></msub><mspace width="1.0em"></mspace><mtext mathvariant="normal">indep.</mtext></mtd><mtd columnalign="left" style="text-align: left"><msub><mi>𝐙</mi><mi>i</mi></msub><mo>∼</mo><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>𝛍</mi><mi>i</mi></msub><mo>,</mo><mi>𝚺</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="center" style="text-align: center"></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">observation space </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>Z</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="1.0em"></mspace><mtext mathvariant="normal">indep.</mtext></mtd><mtd columnalign="left" style="text-align: left"><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>Z</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>∼</mo><mi>𝒫</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo>exp</mo><mo stretchy="false" form="prefix">{</mo><msub><mi>Z</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="postfix">}</mo><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{equation} 
  \begin{array}{rcl}
    \text{group structure }   &amp; \mathbf{\mu}_i = \mu_{g_i} &amp; g_i \in \{1, \dots, K\}  \\
    \text{latent space }   &amp; \mathbf{Z}_i \quad \text{indep.} &amp; \mathbf{Z}_i \sim \mathcal{N}({\boldsymbol\mu}_i, \boldsymbol{\Sigma})  &amp; \\
    \text{observation space } &amp; Y_{ij} | Z_{ij} \quad \text{indep.} &amp; Y_{ij} | Z_{ij} \sim \mathcal{P}\left(\exp\{Z_{ij}\}\right)
  \end{array}
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>g</mi><mi>i</mi></msub><annotation encoding="application/x-tex">g_i</annotation></semantics></math>
denotes the group sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
belongs to.</p>
<p>The different parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>𝛍</mi><mi>k</mi></msub><mo>∈</mo><msup><mi>ℝ</mi><mi>p</mi></msup></mrow><annotation encoding="application/x-tex">{\boldsymbol\mu}_k \in\mathbb{R}^p</annotation></semantics></math>
corresponds to the group-specific main effects and the variance matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝚺</mi><annotation encoding="application/x-tex">\boldsymbol{\Sigma}</annotation></semantics></math>
is shared among groups. An equivalent way of writing this model is the
following:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">latent space </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>𝐙</mi><mi>i</mi></msub><mo>∼</mo><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>𝛍</mi><mi>i</mi></msub><mo>,</mo><mi>𝚺</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><msub><mi>𝛍</mi><mi>i</mi></msub><mo>=</mo><msubsup><mi>𝐠</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝐌</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">observation space </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="center" style="text-align: center"><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>Z</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="1.0em"></mspace><mtext mathvariant="normal">indep.</mtext></mtd><mtd columnalign="left" style="text-align: left"><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>Z</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>∼</mo><mi>𝒫</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo>exp</mo><mo stretchy="false" form="prefix">{</mo><msub><mi>Z</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="false" form="postfix">}</mo><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{equation} 
  \begin{array}{rcl}
  \text{latent space } &amp;   \mathbf{Z}_i \sim \mathcal{N}({\boldsymbol\mu}_i,\boldsymbol\Sigma) &amp; \boldsymbol{\mu}_i = \mathbf{g}_i^\top \mathbf{M} \\
  \text{observation space } &amp;  Y_{ij} | Z_{ij} \quad \text{indep.} &amp; Y_{ij} | Z_{ij} \sim \mathcal{P}\left(\exp\{Z_{ij}\}\right),
  \end{array}
\end{equation}</annotation></semantics></math> where, with a slight
abuse of notation,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐠</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{g}_i</annotation></semantics></math>
is a group-indicator vector of length
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>g</mi><mrow><mi>i</mi><mi>k</mi></mrow></msub><mo>=</mo><mn>1</mn><mo>⇔</mo><msub><mi>g</mi><mi>i</mi></msub><mo>=</mo><mi>k</mi></mrow><annotation encoding="application/x-tex">g_{ik} = 1 \Leftrightarrow g_i = k</annotation></semantics></math>)
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝐌</mi><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>𝛍</mi><mn>1</mn><mi>⊤</mi></msubsup><mo>,</mo><mi>…</mi><mo>,</mo><msubsup><mi>𝛍</mi><mi>K</mi><mi>⊤</mi></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mi>⊤</mi></msup></mrow><annotation encoding="application/x-tex">\mathbf{M} = [\boldsymbol{\mu}_1^\top, \dots, \boldsymbol{\mu}_K^\top]^\top</annotation></semantics></math>
is a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">K \times p</annotation></semantics></math>
matrix collecting the group-specific main effects.</p>
<div class="section level4">
<h4 id="covariates-and-offsets">Covariates and offsets<a class="anchor" aria-label="anchor" href="#covariates-and-offsets"></a>
</h4>
<p>Just like PLN, PLN-LDA generalizes to a formulation close to a
multivariate generalized linear model where the main effect is due to a
linear combination of the discrete group structure,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>
covariates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐱</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{x}_i</annotation></semantics></math>
and a vector
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐨</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{o}_i</annotation></semantics></math>
of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
offsets in sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
The latent layer then reads
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>𝐙</mi><mi>i</mi></msub><mo>∼</mo><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mrow><msub><mi>𝐨</mi><mi>i</mi></msub><mo>+</mo><msubsup><mi>𝐠</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝐌</mi><mo>+</mo><msubsup><mi>𝐱</mi><mi>i</mi><mi>⊤</mi></msubsup><mi>𝐁</mi></mrow><mo>,</mo><mi>𝚺</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\begin{equation} 
  \mathbf{Z}_i \sim \mathcal{N}({\mathbf{o}_i + \mathbf{g}_i^\top \mathbf{M} + \mathbf{x}_i^\top\mathbf{B}},\boldsymbol\Sigma)
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐁</mi><annotation encoding="application/x-tex">\mathbf{B}</annotation></semantics></math>
is a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">d\times p</annotation></semantics></math>
matrix of regression parameters.</p>
</div>
<div class="section level4">
<h4 id="prediction">Prediction<a class="anchor" aria-label="anchor" href="#prediction"></a>
</h4>
<p>Given:</p>
<ul>
<li>a new observation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐘</mi><annotation encoding="application/x-tex">\mathbf{Y}</annotation></semantics></math>
with associated offset
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐨</mi><annotation encoding="application/x-tex">\mathbf{o}</annotation></semantics></math>
and covariates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐱</mi><annotation encoding="application/x-tex">\mathbf{x}</annotation></semantics></math>
</li>
<li>a model with estimated parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{\boldsymbol{\Sigma}}</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝐁</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{\mathbf{B}}</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝐌</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{\mathbf{M}}</annotation></semantics></math>
and group counts
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>n</mi><mn>1</mn></msub><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>n</mi><mi>K</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(n_1, \dots, n_K)</annotation></semantics></math>
</li>
</ul>
<p>We can predict the observation’s group using Bayes rule as follows:
for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>∈</mo><mrow><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>K</mi></mrow></mrow><annotation encoding="application/x-tex">k \in {1, \dots, K}</annotation></semantics></math>,
compute
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="false" form="prefix">|</mo><mi>𝐠</mi><mo>=</mo><mi>k</mi><mo>,</mo><mi>𝐨</mi><mo>,</mo><mi>𝐱</mi><mo>,</mo><mover><mi>𝐁</mi><mo accent="true">̂</mo></mover><mo>,</mo><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><msub><mi>𝚽</mi><mrow><mi>P</mi><mi>L</mi><mi>N</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo>;</mo><mi>𝐨</mi><mo>+</mo><msub><mi>𝛍</mi><mi>k</mi></msub><mo>+</mo><msup><mi>𝐱</mi><mi>⊤</mi></msup><mover><mi>𝐁</mi><mo accent="true">̂</mo></mover><mo>,</mo><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>p</mi><mi>k</mi></msub></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><msub><mi>n</mi><mi>k</mi></msub><mrow><munderover><mo>∑</mo><mrow><mi>k</mi><mi>′</mi><mo>=</mo><mn>1</mn></mrow><mi>K</mi></munderover><msub><mi>n</mi><mrow><mi>k</mi><mi>′</mi></mrow></msub></mrow></mfrac></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{equation}
\begin{aligned}
f_k(\mathbf{Y}) &amp; = p(\mathbf{Y} | \mathbf{g} = k, \mathbf{o}, \mathbf{x}, \hat{\mathbf{B}}, \hat{\boldsymbol{\Sigma}}) \\
                &amp; = \boldsymbol{\Phi}_{PLN}(\mathbf{Y}; \mathbf{o} + \boldsymbol{\mu}_k + \mathbf{x}^\top \hat{\mathbf{B}}, \hat{\boldsymbol{\Sigma}}) \\
p_k             &amp; = \frac{n_k}{\sum_{k' = 1}^K n_{k'}}    
\end{aligned}
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>𝚽</mi><mrow><mi>P</mi><mi>L</mi><mi>N</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mo>•</mo><mo>;</mo><mi>𝛍</mi><mo>,</mo><mi>𝚺</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\boldsymbol{\Phi}_{PLN}(\bullet; \boldsymbol{\mu}, \boldsymbol{\Sigma})</annotation></semantics></math>
is the density function of a PLN distribution with parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝛍</mi><mo>,</mo><mi>𝚺</mi><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(\boldsymbol{\mu}, \boldsymbol{\Sigma})</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">f_k(\mathbf{Y})</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mi>k</mi></msub><annotation encoding="application/x-tex">p_k</annotation></semantics></math>
are respectively plug-in estimates of (i) the probability of observing
counts
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐘</mi><annotation encoding="application/x-tex">\mathbf{Y}</annotation></semantics></math>
in a sample from group
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
and (ii) the probability that a sample originates from group
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>.</p>
<p>The posterior probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\hat{\pi}_k(\mathbf{Y})</annotation></semantics></math>
that observation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐘</mi><annotation encoding="application/x-tex">\mathbf{Y}</annotation></semantics></math>
belongs to group
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
and most likely group
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>k</mi><mo accent="true">̂</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\hat{k}(\mathbf{Y})</annotation></semantics></math>
can thus be defined as
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mfrac><mrow><msub><mi>p</mi><mi>k</mi></msub><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munderover><mo>∑</mo><mrow><mi>k</mi><mi>′</mi><mo>=</mo><mn>1</mn></mrow><mi>K</mi></munderover><msub><mi>p</mi><mrow><mi>k</mi><mi>′</mi></mrow></msub><msub><mi>f</mi><mrow><mi>k</mi><mi>′</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mover><mi>k</mi><mo accent="true">̂</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><munder><mrow><mo>arg</mo><mo>max</mo></mrow><mrow><mi>k</mi><mo>∈</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>K</mi><mo stretchy="false" form="postfix">}</mo></mrow></munder><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝐘</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{equation}
\begin{aligned}
\hat{\pi}_k(\mathbf{Y}) &amp; = \frac{p_k f_k(\mathbf{Y})}{\sum_{k' = 1}^K p_{k'} f_{k'}(\mathbf{Y})} \\
\hat{k}(\mathbf{Y})     &amp; = \underset{k \in \{1, \dots, K\}}{\arg\max} \hat{\pi}_k(\mathbf{Y})
\end{aligned}
\end{equation}</annotation></semantics></math></p>
</div>
<div class="section level4">
<h4 id="optimization-by-variational-inference">Optimization by Variational inference<a class="anchor" aria-label="anchor" href="#optimization-by-variational-inference"></a>
</h4>
<p>Classification and prediction are the main objectives in (PLN-)LDA.
To reach this goal, we first need to estimate the model parameters.
Inference in PLN-LDA focuses on the group-specific main effects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐌</mi><annotation encoding="application/x-tex">\mathbf{M}</annotation></semantics></math>,
the regression parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐁</mi><annotation encoding="application/x-tex">\mathbf{B}</annotation></semantics></math>
and the covariance matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝚺</mi><annotation encoding="application/x-tex">\boldsymbol\Sigma</annotation></semantics></math>.
Technically speaking, we can treat
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>𝐠</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{g}_i</annotation></semantics></math>
as a discrete covariate and estimate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><mi>𝐌</mi><mo>,</mo><mi>𝐁</mi><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[\mathbf{M}, \mathbf{B}]</annotation></semantics></math>
using the same strategy as for the standard <strong>PLN</strong> model.
Briefly, we adopt a variational strategy to approximate the
log-likelihood function and optimize the consecutive variational
surrogate of the log-likelihood with a gradient-ascent-based approach.
To this end, we rely on the CCSA algorithm of <span class="citation">Svanberg (<a href="#ref-Svan02">2002</a>)</span>
implemented in the C++ library <span class="citation">(<a href="#ref-nlopt">Johnson 2011</a>)</span>, which we link to the
package.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="analysis-of-trichoptera-data-with-a-pln-lda-model">Analysis of trichoptera data with a PLN-LDA model<a class="anchor" aria-label="anchor" href="#analysis-of-trichoptera-data-with-a-pln-lda-model"></a>
</h2>
<p>In the package, the PLN-LDA model is adjusted with the function
<code>PLNLDA</code>, which we review in this section. This function
adjusts the model and stores it in a object of class
<code>PLNLDAfit</code> which inherits from the class
<code>PLNfit</code>, so we strongly recommend the reader to be somehow
comfortable with <code>PLN</code> and <code>PLNfit</code> before using
<code>PLNLDA</code> (see <a href="PLN.html">the PLN vignette</a>).</p>
<div class="section level3">
<h3 id="a-model-with-main-effects-and-no-covariates">A model with main effects and no covariates<a class="anchor" aria-label="anchor" href="#a-model-with-main-effects-and-no-covariates"></a>
</h3>
<p>We start by adjusting the above model to the Trichoptera data set. We
use <code>Group</code>, the catching time spans, as a discrete structure
and use log as an offset to capture differences in sampling luck.</p>
<p>The model can be fitted with the function <code>PLNLDA</code> as
follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_nocov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PLNLDA.html">PLNLDA</a></span><span class="op">(</span><span class="va">Abundance</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/offset.html" class="external-link">offset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Offset</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      grouping <span class="op">=</span> <span class="va">Group</span>, </span>
<span>                      data <span class="op">=</span> <span class="va">trichoptera</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Performing discriminant Analysis...</span></span>
<span><span class="co">##  DONE!</span></span></code></pre>
<p>Note that <code>PLNLDA</code> uses the standard <code>formula</code>
interface, like every other model in the <strong>PLNmodels</strong>
package.</p>
<div class="section level4">
<h4 id="structure-of-plnldafit">Structure of <code>PLNLDAfit</code><a class="anchor" aria-label="anchor" href="#structure-of-plnldafit"></a>
</h4>
<p>The <code>myLDA_nocov</code> variable is an <code>R6</code> object
with class <code>PLNLDAfit</code>, which comes with a couple of methods.
The most basic is the <code>show/print</code> method, which sends a
brief summary of the estimation process and available methods:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_nocov</span></span></code></pre></div>
<pre><code><span><span class="co">## Linear Discriminant Analysis for Poisson Lognormal distribution</span></span>
<span><span class="co">## ==================================================================</span></span>
<span><span class="co">##  nb_param   loglik       BIC       ICL</span></span>
<span><span class="co">##       357 -799.746 -1494.436 -1047.782</span></span>
<span><span class="co">## ==================================================================</span></span>
<span><span class="co">## * Useful fields</span></span>
<span><span class="co">##     $model_par, $latent, $latent_pos, $var_par, $optim_par</span></span>
<span><span class="co">##     $loglik, $BIC, $ICL, $loglik_vec, $nb_param, $criteria</span></span>
<span><span class="co">## * Useful S3 methods</span></span>
<span><span class="co">##     print(), coef(), sigma(), vcov(), fitted()</span></span>
<span><span class="co">##     predict(), predict_cond(), standard_error()</span></span>
<span><span class="co">## * Additional fields for LDA</span></span>
<span><span class="co">##     $percent_var, $corr_map, $scores, $group_means</span></span>
<span><span class="co">## * Additional S3 methods for LDA</span></span>
<span><span class="co">##     plot.PLNLDAfit(), predict.PLNLDAfit()</span></span></code></pre>
<p>Comprehensive information about <code>PLNLDAfit</code> is available
via <code><a href="../reference/PLNLDAfit.html">?PLNLDAfit</a></code>.</p>
</div>
<div class="section level4">
<h4 id="specific-fields">Specific fields<a class="anchor" aria-label="anchor" href="#specific-fields"></a>
</h4>
<p>The user can easily access several fields of the
<code>PLNLDAfit</code> object using <code>S3</code> methods, the most
interesting ones are</p>
<ul>
<li>the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">p \times p</annotation></semantics></math>
covariance matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝚺</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{\boldsymbol{\Sigma}}</annotation></semantics></math>:</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/sigma.html" class="external-link">sigma</a></span><span class="op">(</span><span class="va">myLDA_nocov</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">corrplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html" class="external-link">corrplot</a></span><span class="op">(</span>is.corr <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="PLNLDA_files/figure-html/vcov-1.png" width="700"></p>
<ul>
<li>the regression coefficient matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝐁</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{\mathbf{B}}</annotation></semantics></math>
(in this case <code>NULL</code> as there are no covariates)</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">myLDA_nocov</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<ul>
<li>the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>×</mo><mi>K</mi></mrow><annotation encoding="application/x-tex">p \times K</annotation></semantics></math>
matrix of group means
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐌</mi><annotation encoding="application/x-tex">\mathbf{M}</annotation></semantics></math>
</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_nocov</span><span class="op">$</span><span class="va">group_means</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="right">grouping1</th>
<th align="right">grouping2</th>
<th align="right">grouping3</th>
<th align="right">grouping4</th>
<th align="right">grouping5</th>
<th align="right">grouping6</th>
<th align="right">grouping7</th>
<th align="right">grouping8</th>
<th align="right">grouping9</th>
<th align="right">grouping10</th>
<th align="right">grouping11</th>
<th align="right">grouping12</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">-26.81</td>
<td align="right">-6.82</td>
<td align="right">-26.37</td>
<td align="right">-25.20</td>
<td align="right">-28.15</td>
<td align="right">-28.73</td>
<td align="right">-27.04</td>
<td align="right">-26.19</td>
<td align="right">-26.26</td>
<td align="right">-5.64</td>
<td align="right">-3.46</td>
<td align="right">-25.19</td>
</tr>
<tr class="even">
<td align="right">-26.79</td>
<td align="right">-27.53</td>
<td align="right">-5.66</td>
<td align="right">-25.18</td>
<td align="right">-7.44</td>
<td align="right">-28.71</td>
<td align="right">-27.02</td>
<td align="right">-26.17</td>
<td align="right">-26.24</td>
<td align="right">-26.31</td>
<td align="right">-24.15</td>
<td align="right">-4.47</td>
</tr>
<tr class="odd">
<td align="right">-2.38</td>
<td align="right">-3.96</td>
<td align="right">-2.36</td>
<td align="right">-2.92</td>
<td align="right">-6.51</td>
<td align="right">-5.54</td>
<td align="right">-2.69</td>
<td align="right">-2.03</td>
<td align="right">-2.71</td>
<td align="right">-2.65</td>
<td align="right">-24.26</td>
<td align="right">-3.85</td>
</tr>
<tr class="even">
<td align="right">-26.81</td>
<td align="right">-27.56</td>
<td align="right">-5.69</td>
<td align="right">-4.52</td>
<td align="right">-28.09</td>
<td align="right">-28.78</td>
<td align="right">-27.06</td>
<td align="right">-5.51</td>
<td align="right">-5.59</td>
<td align="right">-4.91</td>
<td align="right">-24.21</td>
<td align="right">-25.23</td>
</tr>
<tr class="odd">
<td align="right">-0.28</td>
<td align="right">-0.26</td>
<td align="right">-0.62</td>
<td align="right">-1.09</td>
<td align="right">-0.61</td>
<td align="right">-0.11</td>
<td align="right">-0.66</td>
<td align="right">-0.80</td>
<td align="right">-0.39</td>
<td align="right">-0.45</td>
<td align="right">-0.61</td>
<td align="right">-1.16</td>
</tr>
<tr class="even">
<td align="right">-4.00</td>
<td align="right">-3.32</td>
<td align="right">-2.93</td>
<td align="right">-4.47</td>
<td align="right">-6.72</td>
<td align="right">-8.00</td>
<td align="right">-2.85</td>
<td align="right">-3.06</td>
<td align="right">-5.53</td>
<td align="right">-3.99</td>
<td align="right">-24.13</td>
<td align="right">-25.15</td>
</tr>
</tbody>
</table>
<p>The <code>PLNLDAfit</code> class also benefits from two important
methods: <code>plot</code> and <code>predict</code>.</p>
</div>
<div class="section level4">
<h4 id="plot-method">
<code>plot</code> method<a class="anchor" aria-label="anchor" href="#plot-method"></a>
</h4>
<p>The <code>plot</code> methods provides easy to interpret graphics
which reveals here that the groups are well separated:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">myLDA_nocov</span><span class="op">)</span></span></code></pre></div>
<p><img src="PLNLDA_files/figure-html/plot_model-1.png" width="672"></p>
<p>By default, <code>plot</code> shows the first 3 axis of the LDA when
there are 4 or more groups and uses special representations for the edge
cases of 3 or less groups.</p>
<p><code>ggplot2</code>-savvy users who want to make their own
representations can extracts the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>K</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">n \times (K-1)</annotation></semantics></math>
matrix of sample scores from the <code>PLNLDAfit</code> object …</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_nocov</span><span class="op">$</span><span class="va">scores</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="8%">
<col width="9%">
<col width="10%">
<col width="9%">
<col width="10%">
<col width="9%">
<col width="8%">
<col width="9%">
<col width="7%">
<col width="6%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="right">LD1</th>
<th align="right">LD2</th>
<th align="right">LD3</th>
<th align="right">LD4</th>
<th align="right">LD5</th>
<th align="right">LD6</th>
<th align="right">LD7</th>
<th align="right">LD8</th>
<th align="right">LD9</th>
<th align="right">LD10</th>
<th align="right">LD11</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">5572851</td>
<td align="right">-1314725</td>
<td align="right">-13445.07</td>
<td align="right">108475.9</td>
<td align="right">-32807.19</td>
<td align="right">13932.96</td>
<td align="right">2222.78</td>
<td align="right">-1121.16</td>
<td align="right">127.57</td>
<td align="right">63.41</td>
<td align="right">-162.06</td>
</tr>
<tr class="even">
<td align="right">5577052</td>
<td align="right">-1309516</td>
<td align="right">-13663.61</td>
<td align="right">109043.8</td>
<td align="right">-32620.18</td>
<td align="right">13944.36</td>
<td align="right">2245.39</td>
<td align="right">-1159.62</td>
<td align="right">122.38</td>
<td align="right">65.72</td>
<td align="right">-163.25</td>
</tr>
<tr class="odd">
<td align="right">5615995</td>
<td align="right">-1262057</td>
<td align="right">-16024.20</td>
<td align="right">112918.3</td>
<td align="right">-31001.54</td>
<td align="right">13645.60</td>
<td align="right">1928.42</td>
<td align="right">-1149.70</td>
<td align="right">157.40</td>
<td align="right">70.11</td>
<td align="right">-156.01</td>
</tr>
<tr class="even">
<td align="right">5659650</td>
<td align="right">-1208386</td>
<td align="right">-18638.87</td>
<td align="right">117093.7</td>
<td align="right">-29209.81</td>
<td align="right">13210.21</td>
<td align="right">1415.06</td>
<td align="right">-1006.97</td>
<td align="right">207.65</td>
<td align="right">71.59</td>
<td align="right">-146.86</td>
</tr>
<tr class="odd">
<td align="right">5634198</td>
<td align="right">-1239487</td>
<td align="right">-17081.41</td>
<td align="right">114602.3</td>
<td align="right">-30272.75</td>
<td align="right">13417.37</td>
<td align="right">1648.60</td>
<td align="right">-1028.38</td>
<td align="right">182.73</td>
<td align="right">69.74</td>
<td align="right">-151.75</td>
</tr>
<tr class="even">
<td align="right">5602064</td>
<td align="right">-1278862</td>
<td align="right">-15153.85</td>
<td align="right">111489.5</td>
<td align="right">-31589.74</td>
<td align="right">13712.83</td>
<td align="right">1980.78</td>
<td align="right">-1101.02</td>
<td align="right">149.14</td>
<td align="right">67.21</td>
<td align="right">-158.22</td>
</tr>
</tbody>
</table>
<p>…or the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>×</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>K</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">p \times (K-1)</annotation></semantics></math>
matrix of correlations between scores and (latent) variables</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_nocov</span><span class="op">$</span><span class="va">corr_map</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="5%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">LD1</th>
<th align="right">LD2</th>
<th align="right">LD3</th>
<th align="right">LD4</th>
<th align="right">LD5</th>
<th align="right">LD6</th>
<th align="right">LD7</th>
<th align="right">LD8</th>
<th align="right">LD9</th>
<th align="right">LD10</th>
<th align="right">LD11</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Che</td>
<td align="right">-0.31</td>
<td align="right">-0.09</td>
<td align="right">-0.52</td>
<td align="right">-0.25</td>
<td align="right">0.63</td>
<td align="right">-0.26</td>
<td align="right">0.27</td>
<td align="right">-0.41</td>
<td align="right">0.85</td>
<td align="right">-0.47</td>
<td align="right">0.02</td>
</tr>
<tr class="even">
<td align="left">Hyc</td>
<td align="right">0.07</td>
<td align="right">0.58</td>
<td align="right">-0.01</td>
<td align="right">-0.39</td>
<td align="right">0.00</td>
<td align="right">-0.56</td>
<td align="right">-0.76</td>
<td align="right">0.75</td>
<td align="right">-0.08</td>
<td align="right">-0.24</td>
<td align="right">0.57</td>
</tr>
<tr class="odd">
<td align="left">Hym</td>
<td align="right">0.11</td>
<td align="right">0.18</td>
<td align="right">0.11</td>
<td align="right">0.74</td>
<td align="right">-0.27</td>
<td align="right">-0.08</td>
<td align="right">0.37</td>
<td align="right">0.23</td>
<td align="right">-0.35</td>
<td align="right">0.05</td>
<td align="right">-0.60</td>
</tr>
<tr class="even">
<td align="left">Hys</td>
<td align="right">-0.45</td>
<td align="right">0.16</td>
<td align="right">0.58</td>
<td align="right">0.02</td>
<td align="right">-0.21</td>
<td align="right">-0.36</td>
<td align="right">0.45</td>
<td align="right">0.25</td>
<td align="right">-0.32</td>
<td align="right">-0.06</td>
<td align="right">0.25</td>
</tr>
<tr class="odd">
<td align="left">Psy</td>
<td align="right">0.14</td>
<td align="right">-0.48</td>
<td align="right">-0.24</td>
<td align="right">0.28</td>
<td align="right">0.39</td>
<td align="right">0.26</td>
<td align="right">0.09</td>
<td align="right">-0.03</td>
<td align="right">0.48</td>
<td align="right">-0.55</td>
<td align="right">-0.38</td>
</tr>
<tr class="even">
<td align="left">Aga</td>
<td align="right">0.13</td>
<td align="right">0.11</td>
<td align="right">0.17</td>
<td align="right">0.88</td>
<td align="right">-0.03</td>
<td align="right">-0.16</td>
<td align="right">0.26</td>
<td align="right">-0.06</td>
<td align="right">-0.13</td>
<td align="right">-0.17</td>
<td align="right">-0.52</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="predict-method">
<code>predict</code> method<a class="anchor" aria-label="anchor" href="#predict-method"></a>
</h4>
<p>The <code>predict</code> method has a slightly different behavior
than its siblings in other models of the <strong>PLNmodels</strong>. The
goal of <code>predict</code> is to predict the discrete class based on
observed <em>species counts</em> (rather than predicting counts from
known covariates).</p>
<p>By default, the <code>predict</code> use the argument
<code>type = "posterior"</code> to output the matrix of log-posterior
probabilities
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>π</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mi>k</mi></msub></mrow><annotation encoding="application/x-tex">\log(\hat{\pi})_k</annotation></semantics></math></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">predicted.class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">myLDA_nocov</span>, newdata <span class="op">=</span> <span class="va">trichoptera</span><span class="op">)</span></span>
<span><span class="co">## equivalent to </span></span>
<span><span class="co">## predicted.class &lt;- predict(myLDA_nocov, newdata = trichoptera,  type = "posterior")</span></span>
<span><span class="va">predicted.class</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="7%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="7%">
<col width="7%">
<col width="8%">
<col width="8%">
<col width="7%">
<col width="8%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="right">1</th>
<th align="right">2</th>
<th align="right">3</th>
<th align="right">4</th>
<th align="right">5</th>
<th align="right">6</th>
<th align="right">7</th>
<th align="right">8</th>
<th align="right">9</th>
<th align="right">10</th>
<th align="right">11</th>
<th align="right">12</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">-17.45</td>
<td align="right">-46.31</td>
<td align="right">-43.36</td>
<td align="right">-25.48</td>
<td align="right">-54.51</td>
<td align="right">-32.75</td>
<td align="right">-25.92</td>
<td align="right">-86.97</td>
<td align="right">-84.04</td>
<td align="right">-20.60</td>
<td align="right">-206.17</td>
<td align="right">-74.51</td>
</tr>
<tr class="even">
<td align="right">-9.67</td>
<td align="right">-17.04</td>
<td align="right">-59.59</td>
<td align="right">-21.33</td>
<td align="right">-71.84</td>
<td align="right">-27.30</td>
<td align="right">-14.69</td>
<td align="right">-15.58</td>
<td align="right">-15.21</td>
<td align="right">-14.70</td>
<td align="right">-121.03</td>
<td align="right">-66.07</td>
</tr>
<tr class="odd">
<td align="right">-12.94</td>
<td align="right">-22.57</td>
<td align="right">-114.44</td>
<td align="right">-39.98</td>
<td align="right">-119.29</td>
<td align="right">-30.79</td>
<td align="right">-22.53</td>
<td align="right">-49.90</td>
<td align="right">-44.31</td>
<td align="right">-22.94</td>
<td align="right">-148.03</td>
<td align="right">-119.90</td>
</tr>
<tr class="even">
<td align="right">-18.43</td>
<td align="right">-31.01</td>
<td align="right">-121.46</td>
<td align="right">-116.12</td>
<td align="right">-143.58</td>
<td align="right">-54.04</td>
<td align="right">-41.91</td>
<td align="right">-114.00</td>
<td align="right">-102.71</td>
<td align="right">-44.65</td>
<td align="right">-311.44</td>
<td align="right">-226.39</td>
</tr>
<tr class="odd">
<td align="right">-13.84</td>
<td align="right">-20.00</td>
<td align="right">-52.55</td>
<td align="right">-60.96</td>
<td align="right">-72.50</td>
<td align="right">-38.18</td>
<td align="right">-26.89</td>
<td align="right">-51.63</td>
<td align="right">-45.10</td>
<td align="right">-25.88</td>
<td align="right">-195.26</td>
<td align="right">-115.17</td>
</tr>
<tr class="even">
<td align="right">-9.25</td>
<td align="right">-13.48</td>
<td align="right">-37.05</td>
<td align="right">-24.40</td>
<td align="right">-48.01</td>
<td align="right">-23.74</td>
<td align="right">-14.61</td>
<td align="right">-15.91</td>
<td align="right">-15.30</td>
<td align="right">-14.32</td>
<td align="right">-100.40</td>
<td align="right">-67.41</td>
</tr>
</tbody>
</table>
<p>You can also show them in the standard (and human-friendly)
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[0, 1]</annotation></semantics></math>
scale with <code>scale = "prob"</code> to get the matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mi>k</mi></msub><annotation encoding="application/x-tex">\hat{\pi}_k</annotation></semantics></math></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">predicted.class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">myLDA_nocov</span>, newdata <span class="op">=</span> <span class="va">trichoptera</span>, scale <span class="op">=</span> <span class="st">"prob"</span><span class="op">)</span></span>
<span><span class="va">predicted.class</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">1</th>
<th align="right">2</th>
<th align="right">3</th>
<th align="right">4</th>
<th align="right">5</th>
<th align="right">6</th>
<th align="right">7</th>
<th align="right">8</th>
<th align="right">9</th>
<th align="right">10</th>
<th align="right">11</th>
<th align="right">12</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">0.958</td>
<td align="right">0.000</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.041</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">0.980</td>
<td align="right">0.001</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.006</td>
<td align="right">0.003</td>
<td align="right">0.004</td>
<td align="right">0.006</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="right">1.000</td>
<td align="right">0.000</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">1.000</td>
<td align="right">0.000</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="right">0.998</td>
<td align="right">0.002</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">0.972</td>
<td align="right">0.014</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.005</td>
<td align="right">0.001</td>
<td align="right">0.002</td>
<td align="right">0.006</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>Setting <code>type = "response"</code>, we can predict the most
likely group
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>k</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{k}</annotation></semantics></math>
instead:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">predicted.class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">myLDA_nocov</span>, newdata <span class="op">=</span> <span class="va">trichoptera</span>,  type <span class="op">=</span> <span class="st">"response"</span><span class="op">)</span></span>
<span><span class="va">predicted.class</span></span></code></pre></div>
<pre><code><span><span class="co">##  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 </span></span>
<span><span class="co">##  1  1  1  1  1  1  9  2  1  1  1  1  2  3  2  2  2  3  3  3  9  3  4  1  4  4 </span></span>
<span><span class="co">## 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 </span></span>
<span><span class="co">## 12  5  4  5  6  7  7  7  8  8  8  8  8  1  9  9  9 10 10 10 10 11 12 </span></span>
<span><span class="co">## Levels: 1 2 3 4 5 6 7 8 9 10 11 12</span></span></code></pre>
<p>We can assess that the predictions are quite similar to the real
group (<em>this is not a proper validation of the method as we used data
set for both model fitting and prediction and are thus at risk of
overfitting</em>).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">predicted.class</span>, <span class="va">trichoptera</span><span class="op">$</span><span class="va">Group</span>, dnn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"predicted"</span>, <span class="st">"true"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          true</span></span>
<span><span class="co">## predicted  1  2  3  4  5  6  7  8  9 10 11 12</span></span>
<span><span class="co">##        1  10  0  0  1  0  0  0  0  1  0  0  0</span></span>
<span><span class="co">##        2   1  4  0  0  0  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        3   0  1  4  0  0  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        4   0  0  0  3  1  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        5   0  0  0  0  2  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        6   0  0  0  0  0  1  0  0  0  0  0  0</span></span>
<span><span class="co">##        7   0  0  0  0  0  0  3  0  0  0  0  0</span></span>
<span><span class="co">##        8   0  0  0  0  0  0  0  4  1  0  0  0</span></span>
<span><span class="co">##        9   1  0  1  0  0  0  0  0  3  0  0  0</span></span>
<span><span class="co">##        10  0  0  0  0  0  0  0  0  0  4  0  0</span></span>
<span><span class="co">##        11  0  0  0  0  0  0  0  0  0  0  1  0</span></span>
<span><span class="co">##        12  0  0  0  0  1  0  0  0  0  0  0  1</span></span></code></pre>
<p>Finally, we can get the coordinates of the new data on the same graph
at the original ones with <code>type = "scores"</code>. This is done by
averaging the latent positions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>𝐙</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo>+</mo><msub><mi>𝛍</mi><mi>k</mi></msub></mrow><annotation encoding="application/x-tex">\hat{\mathbf{Z}}_i + \boldsymbol{\mu}_k</annotation></semantics></math>
(found when the sample is assumed to come from group
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>)
and weighting them with the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mover><mi>π</mi><mo accent="true">̂</mo></mover><mi>k</mi></msub><annotation encoding="application/x-tex">\hat{\pi}_k</annotation></semantics></math>.
Some samples, have compositions that put them very far from their group
mean.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">predicted.scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">myLDA_nocov</span>, newdata <span class="op">=</span> <span class="va">trichoptera</span>,  type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">predicted.scores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Axis."</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">predicted.scores</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">predicted.scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">predicted.scores</span><span class="op">)</span></span>
<span><span class="va">predicted.scores</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">trichoptera</span><span class="op">$</span><span class="va">Group</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">myLDA_nocov</span>, map <span class="op">=</span> <span class="st">"individual"</span>, nb_axes <span class="op">=</span> <span class="fl">2</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">predicted.scores</span>, </span>
<span>             <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Axis.1</span>, y <span class="op">=</span> <span class="va">Axis.2</span>, color <span class="op">=</span> <span class="va">group</span>, label <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="PLNLDA_files/figure-html/predicted_scores-1.png" width="672"></p>
</div>
</div>
<div class="section level3">
<h3 id="a-model-with-latent-main-effects-and-meteorological-covariates">A model with latent main effects and meteorological covariates<a class="anchor" aria-label="anchor" href="#a-model-with-latent-main-effects-and-meteorological-covariates"></a>
</h3>
<p>It is possible to correct for other covariates before finding the LDA
axes that best separate well the groups. In our case ,we’re going to use
<code>Wind</code> as a covariate and illustrate the main differences
with before :</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PLNLDA.html">PLNLDA</a></span><span class="op">(</span><span class="va">Abundance</span> <span class="op">~</span> <span class="va">Wind</span> <span class="op">+</span> <span class="fl">0</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/offset.html" class="external-link">offset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Offset</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                    grouping <span class="op">=</span> <span class="va">Group</span>, </span>
<span>                    data <span class="op">=</span> <span class="va">trichoptera</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Performing discriminant Analysis...</span></span>
<span><span class="co">##  DONE!</span></span></code></pre>
<div class="section level4">
<h4 id="specific-fields-1">Specific fields<a class="anchor" aria-label="anchor" href="#specific-fields-1"></a>
</h4>
<p>All fields of our new <code>PLNLDA</code> fit can be accessed as
before with similar results. The only important difference is the result
of <code>coef</code>: since we included a covariate in the model,
<code>coef</code> now returns a 1-column matrix for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>𝐁</mi><mo accent="true">̂</mo></mover><annotation encoding="application/x-tex">\hat{\mathbf{B}}</annotation></semantics></math>
instead of <code>NULL</code></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">myLDA_cov</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Wind</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Che</td>
<td align="right">-0.3292299</td>
</tr>
<tr class="even">
<td align="left">Hyc</td>
<td align="right">1.3723327</td>
</tr>
<tr class="odd">
<td align="left">Hym</td>
<td align="right">-0.1927340</td>
</tr>
<tr class="even">
<td align="left">Hys</td>
<td align="right">-0.4544647</td>
</tr>
<tr class="odd">
<td align="left">Psy</td>
<td align="right">0.0372515</td>
</tr>
<tr class="even">
<td align="left">Aga</td>
<td align="right">-0.0611397</td>
</tr>
</tbody>
</table>
<p>The group-specific main effects can still be accessed with
<code>$group_means</code></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myLDA_cov</span><span class="op">$</span><span class="va">group_means</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="right">grouping1</th>
<th align="right">grouping2</th>
<th align="right">grouping3</th>
<th align="right">grouping4</th>
<th align="right">grouping5</th>
<th align="right">grouping6</th>
<th align="right">grouping7</th>
<th align="right">grouping8</th>
<th align="right">grouping9</th>
<th align="right">grouping10</th>
<th align="right">grouping11</th>
<th align="right">grouping12</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">-22.67</td>
<td align="right">-11.35</td>
<td align="right">-27.03</td>
<td align="right">-21.43</td>
<td align="right">-28.73</td>
<td align="right">-48.57</td>
<td align="right">-33.33</td>
<td align="right">-31.56</td>
<td align="right">-21.58</td>
<td align="right">5.81</td>
<td align="right">13.54</td>
<td align="right">-26.85</td>
</tr>
<tr class="even">
<td align="right">-26.24</td>
<td align="right">-46.86</td>
<td align="right">-0.70</td>
<td align="right">-19.10</td>
<td align="right">-9.80</td>
<td align="right">-52.78</td>
<td align="right">-36.62</td>
<td align="right">-33.33</td>
<td align="right">-17.65</td>
<td align="right">-20.07</td>
<td align="right">-8.47</td>
<td align="right">2.34</td>
</tr>
<tr class="odd">
<td align="right">-1.00</td>
<td align="right">-4.50</td>
<td align="right">-1.66</td>
<td align="right">-1.61</td>
<td align="right">-5.73</td>
<td align="right">-7.08</td>
<td align="right">-2.65</td>
<td align="right">-1.81</td>
<td align="right">-1.37</td>
<td align="right">-1.37</td>
<td align="right">-29.03</td>
<td align="right">-3.22</td>
</tr>
<tr class="even">
<td align="right">-27.64</td>
<td align="right">-36.74</td>
<td align="right">-2.49</td>
<td align="right">0.95</td>
<td align="right">-30.64</td>
<td align="right">-42.33</td>
<td align="right">-33.58</td>
<td align="right">-4.99</td>
<td align="right">0.32</td>
<td align="right">0.91</td>
<td align="right">-22.10</td>
<td align="right">-29.30</td>
</tr>
<tr class="odd">
<td align="right">-0.05</td>
<td align="right">-0.43</td>
<td align="right">-0.47</td>
<td align="right">-0.85</td>
<td align="right">-0.44</td>
<td align="right">-0.43</td>
<td align="right">-0.59</td>
<td align="right">-0.71</td>
<td align="right">-0.06</td>
<td align="right">-0.15</td>
<td align="right">-0.19</td>
<td align="right">-1.03</td>
</tr>
<tr class="even">
<td align="right">-1.64</td>
<td align="right">-4.18</td>
<td align="right">-1.61</td>
<td align="right">-2.23</td>
<td align="right">-5.41</td>
<td align="right">-10.74</td>
<td align="right">-2.70</td>
<td align="right">-2.71</td>
<td align="right">-3.11</td>
<td align="right">-1.56</td>
<td align="right">-27.37</td>
<td align="right">-30.89</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="plot-method-1">
<code>plot</code> method<a class="anchor" aria-label="anchor" href="#plot-method-1"></a>
</h4>
<p>Once again, the <code>plot</code> method is very useful to get a
quick look at the results.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">myLDA_cov</span><span class="op">)</span></span></code></pre></div>
<p><img src="PLNLDA_files/figure-html/plot_model_cov-1.png" width="672"></p>
</div>
<div class="section level4">
<h4 id="predict-method-1">
<code>predict</code> method<a class="anchor" aria-label="anchor" href="#predict-method-1"></a>
</h4>
<p>We can again predict the most likely group for each sample :</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">predicted.class_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">myLDA_cov</span>, newdata <span class="op">=</span> <span class="va">trichoptera</span>, type <span class="op">=</span> <span class="st">"response"</span><span class="op">)</span></span></code></pre></div>
<p>and check that we recover the correct class in most cases (again, we
used the same data set for model fitting and group prediction only for
ease of exposition):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">predicted.class_cov</span>, <span class="va">trichoptera</span><span class="op">$</span><span class="va">Group</span>, dnn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"predicted"</span>, <span class="st">"true"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          true</span></span>
<span><span class="co">## predicted  1  2  3  4  5  6  7  8  9 10 11 12</span></span>
<span><span class="co">##        1  11  0  0  1  0  0  0  0  1  0  0  0</span></span>
<span><span class="co">##        2   1  4  0  0  0  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        3   0  1  4  0  0  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        4   0  0  0  3  1  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        5   0  0  0  0  2  0  0  0  0  0  0  0</span></span>
<span><span class="co">##        6   0  0  0  0  0  1  0  0  0  0  0  0</span></span>
<span><span class="co">##        7   0  0  0  0  0  0  3  0  0  0  0  0</span></span>
<span><span class="co">##        8   0  0  0  0  0  0  0  4  1  0  0  0</span></span>
<span><span class="co">##        9   0  0  1  0  0  0  0  0  3  0  0  0</span></span>
<span><span class="co">##        10  0  0  0  0  0  0  0  0  0  4  0  0</span></span>
<span><span class="co">##        11  0  0  0  0  0  0  0  0  0  0  1  0</span></span>
<span><span class="co">##        12  0  0  0  0  1  0  0  0  0  0  0  1</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-AiH89" class="csl-entry">
Aitchison, J., and C. H. Ho. 1989. <span>“The Multivariate Poisson-Log
Normal Distribution.”</span> <em>Biometrika</em> 76 (4): 643–53.
</div>
<div id="ref-Fi36" class="csl-entry">
Fisher, R. A. 1936. <span>“The Use of Multiple Measurements in Taxonomic
Problems.”</span> <em>Annals of Eugenics</em> 7 (2): 179–88.
</div>
<div id="ref-nlopt" class="csl-entry">
Johnson, Steven G. 2011. <em>The NLopt Nonlinear-Optimization
Package</em>. <a href="https://nlopt.readthedocs.io/en/latest/" class="external-link">https://nlopt.readthedocs.io/en/latest/</a>.
</div>
<div id="ref-Rao48" class="csl-entry">
Rao, C. Radhakrishna. 1948. <span>“The Utilization of Multiple
Measurements in Problems of Biological Classification.”</span>
<em>Journal of the Royal Statistical Society. Series B
(Methodological)</em> 10 (2): 159–203.
</div>
<div id="ref-Svan02" class="csl-entry">
Svanberg, Krister. 2002. <span>“A Class of Globally Convergent
Optimization Methods Based on Conservative Convex Separable
Approximations.”</span> <em>SIAM Journal on Optimization</em> 12 (2):
555–73.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Julien Chiquet, Mahendra Mariadassou, Stéphane Robin, François Gindraud.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
